---
title: "2_DESeq_analysis"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, warning = TRUE, message = FALSE, cache = TRUE)
```

```{bash, engine.opts='-l'}
echo $PATH
```

### Load libraries
```{r load_libraries, inlcude = TRUE}

## clear
rm(list=ls())

## Install Rtools directly from (https://cran.r-project.org/bin/windows/Rtools/), then install these on first run:
# install.packages("BiocManager")
# BiocManager::install("DESeq2")
# BiocManager::install("vsn")
# BiocManager::install("tidybulk")
# BiocManager::install("goseq")
# BiocManager::install("affycoretools")
# BiocManager::install("EnhancedVolcano")
# BiocManager::install("pcaExplorer")
# BiocManager::install("apeglm")
# BiocManager::install("PCAtools")


# List of packages we want to install (run every time)
load.lib<-c("DESeq2","edgeR","goseq","dplyr","GenomicFeatures","data.table","calibrate","affycoretools","data.table","vsn","tidybulk","ggplot2","cowplot","pheatmap","gplots","RColorBrewer","EnhancedVolcano","pcaExplorer","readxl","apeglm","ashr","tibble","plotly","sqldf","PCAtools","ggpubr","beepr","genefilter","ComplexHeatmap","circlize","scales")

# Select only the packages that aren't currently installed (run every time)
install.lib <- load.lib[!load.lib %in% installed.packages()]

# And finally we install the missing packages, including their dependency.
for(lib in install.lib) install.packages(lib,dependencies=TRUE)
# After the installation process completes, we load all packages.
sapply(load.lib,require,character=TRUE)
                        
```

#Set ggplot theme
```{r ggplot_theme, include=FALSE}

my_theme <- theme(line              = element_line(size=1.5),
                  rect              = element_rect(size=1.5),
                  text              = element_text(size=14,color="black"),
                  panel.background  = element_blank(),
                  panel.grid.major  = element_blank(), 
                  panel.grid.minor  = element_blank(),
                  axis.text.x       = element_text(size=16,color="black"),
                  axis.text.y       = element_text(size=16,color="black"),
                  axis.title.x      = element_text(margin = margin(t = 10, r = 0, b = 0, l = 0)),
                  axis.title.y      = element_text(margin = margin(t = 0, r = 10, b = 0, l = 0)),
                  axis.ticks.x      = element_line(color="black"),
                  axis.ticks.y      = element_line(color="black"),
                  # axis.line         = element_line(color = "black", size = 0.1),
                  panel.border      = element_rect(color = "black", fill=NA, size=1.5),
                  legend.key        = element_blank()) # removes background of legend bullets

blank_theme <-  theme_minimal()+
                theme(
                axis.title.x = element_blank(),
                axis.title.y = element_blank(),
                panel.border = element_blank(),
                panel.grid=element_blank(),
                axis.ticks = element_blank(),
                plot.title=element_text(size=14, face="bold")
                )

```

#Load data
```{r load_data, inlcude = TRUE}

# Load LOCID to gene description tables, compare, find gigas and mito genes
a1 <- read.delim("data/GCF_902806645.1_cgigas_uk_roslin_v1_feature_table.txt", header = TRUE)
a1 <- a1 %>% filter(X..feature == "CDS")
a1 <- a1[,c("symbol","name")]
colnames(a1) <- c("LOCID","description")
a1 <- distinct(a1, LOCID, .keep_all = TRUE)
a2 <- read.delim("data/LOCID_38222_Roslin_Gigas_mito.txt", header = TRUE)
colnames(a2) <- c("LOCID","description")
a3 <- anti_join(a2,a1, by="LOCID")
a1$feature <- 'genome'
a3$feature <- 'mito'
LOCID_Roslin_gigas_gene_mito_table <- as.data.frame(rbind(a1,a3))
LOCID_Roslin_gigas_gene_mito_table <- distinct(LOCID_Roslin_gigas_gene_mito_table, LOCID, .keep_all = TRUE)
write.table(LOCID_Roslin_gigas_gene_mito_table, "data/LOCID_Roslin_gigas_gene_mito.txt")
rm(a1,a2,a3)

# Import experimental design information
trt_list <- read.csv("data/treatments.csv", sep=',', header=TRUE, row.names = "ID")
coldata <- trt_list[,c("ploidy","condition")]

# Import gene_count_matrix
cts <- as.matrix(read.csv("data/20220906_HISAT2_gene_count_matrix.csv", sep=',', header=TRUE, row.names = "gene_id"))
colnames(cts) <- row.names(trt_list) #!make sure column and row names for coldata and cts match
all(colnames(cts) %in% rownames(coldata))

# Factor independent variables
coldata$ploidy <- factor(coldata$ploidy)
coldata$condition <- factor(coldata$condition)
coldata$group <- factor(trt_list$group) # for individual pair-wise comparisons

# Remove bad samples
# MuliQC report: D54, N56, X44
# Pheatmap outliers: R53?
# BIPLOT outliers: R53, M43, N54, X42 
coldata <- coldata[!(row.names(coldata) %in% c('D54','N56', 'X44', 'R53', 'M43', 'N54', 'X42','T62')),]
cts <- as.matrix(subset(cts, select=-c(D54, N56, X44, R53, M43, N54, X42, T62)))
coldata %>% dplyr::count(group)
all(colnames(cts) %in% rownames(coldata))

# Reorder cts/coldata by group for easy visualization in pcaexplorer
coldata <- coldata %>% arrange(group)
col.order <- rownames(coldata)
cts <- cts[,col.order]

# Remove gene-LOC designation from LOC #
rs <- (rownames(cts))
rs <- strsplit(rs, split = "LOC")
output <- matrix(ncol=1, nrow=length(rs))
# colnames(output) <- c("x","y","z")
for (i in 1:length(rs)){
  LOC_number <- as.data.frame(rs[[i]])
  add_val <- paste("LOC",LOC_number[3,1],sep="")
  output[i,] <- add_val
}
rownames(cts) <- output

```

# Create output folders
```{bash}
# make output folder
mkdir DESEQ2_output

# samples included _ variable analyzed (when 1 specified, compared to control)
mkdir DESEQ2_output/all_treatments
mkdir DESEQ2_output/control_ploidy
mkdir DESEQ2_output/heat_ploidy
mkdir DESEQ2_output/desiccation_ploidy
mkdir DESEQ2_output/diploid_heat
mkdir DESEQ2_output/diploid_desiccation
mkdir DESEQ2_output/triploid_heat
mkdir DESEQ2_output/triploid_desiccation
mkdir DESEQ2_output/diploid
mkdir DESEQ2_output/triploid

```

# ALL treatments, both ploidy
```{r ALL:TRT:ploidy, warning=FALSE, include=TRUE}

# Filter data
coldata_trt <- coldata # %>% filter(group == "A" | group == "B")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ group)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across 1/3 of the samples - https://support.bioconductor.org/p/110307/
keep <- rowSums(DESeq2::counts(dds) >= 10) >= ncol(cts_trt)/3
dds <- dds[keep,]

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)

bp0 <- biplot(p, showLoadings = FALSE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

bp0

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'group',
              colkey = c('A' = 'royalblue1' , 
                         'B' = 'green1'     ,
                         'C' = 'yellow2' ,
                         'D' = 'orange' ,
                         'E' = 'orangered1',
                         'F' = 'red3'),
              shape = 'ploidy',
              shapekey = c('diploid' = 19,  # circle
                           'triploid' = 17), # triangle
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              # ellipseFillKey = c('diploid' = 'blue', 'triploid' = 'green'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-150,150), ylim = c(-80, 50),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1

bp2 <- pairsplot(p,
                components = getComponents(p, c(1:5)),
                triangle = TRUE, trianglelabSize = 12,
                hline = 0, vline = 0,
                pointSize = 1,
                gridlines.major = FALSE, gridlines.minor = FALSE,
                colby = 'group',
                title = 'Pairs plot', plotaxes = FALSE,
                margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2

# Plot pheatmap
rld <- rlog(dds, blind=T)
rld_mat <- assay(rld)
pca <- prcomp(t(rld_mat))
rld_cor <- cor(rld_mat)
bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])

bp3

setwd("DESEQ2_output/all_treatments")

ggexport(filename = "ALL-TREATMENTS-BIPLOT.png",
         plot   = bp0,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "ALL-TREATMENTS-PCA.png",
         plot   = bp1,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "ALL-TREATMENTS-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

ggexport(filename = "ALL-TREATMENTS-pheatmap.png",
         plot   = bp3,
         res    = 600,
         width  = 6000,
         height = 6000)

# pcaExplorer(dds = dds_ploidy)
beep(2)
```

# ALL treatments, diploids only
```{r ALL:TRT:diploids, warning=FALSE, include=TRUE}

# Filter data
coldata_trt <- coldata %>% filter(group == "A" | group == "C" | group == "E")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ group)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across 1/3 of the samples - https://support.bioconductor.org/p/110307/
keep <- rowSums(DESeq2::counts(dds) >= 10) >= ncol(cts_trt)/3
dds <- dds[keep,]

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)

bp0 <- biplot(p, showLoadings = FALSE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

bp0

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'group',
              colkey = c('A' = 'royalblue1' , 
                         'B' = 'green1'     ,
                         'C' = 'yellow2' ,
                         'D' = 'orange' ,
                         'E' = 'orangered1',
                         'F' = 'red3'),
              shape = 'ploidy',
              shapekey = c('diploid' = 19,  # circle
                           'triploid' = 17), # triangle
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              # ellipseFillKey = c('diploid' = 'blue', 'triploid' = 'green'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-150,140), ylim = c(-60, 40),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1

bp2 <- pairsplot(p,
                components = getComponents(p, c(1:5)),
                triangle = TRUE, trianglelabSize = 12,
                hline = 0, vline = 0,
                pointSize = 1,
                gridlines.major = FALSE, gridlines.minor = FALSE,
                colby = 'group',
                title = 'Pairs plot', plotaxes = FALSE,
                margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2

# Plot pheatmap
rld <- rlog(dds, blind=T)
rld_mat <- assay(rld)
pca <- prcomp(t(rld_mat))
rld_cor <- cor(rld_mat)
bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])

bp3

setwd("DESEQ2_output/diploid")

ggexport(filename = "ALL-TREATMENTS-DIPLOID-BIPLOT.png",
         plot   = bp0,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "ALL-TREATMENTS-DIPLOID-PCA.png",
         plot   = bp1,
         res    = 600,
         # device = "png",
         width  = 4000,
         height = 5000)

ggexport(filename = "ALL-TREATMENTS-DIPLOID-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

ggexport(filename = "ALL-TREATMENTS-DIPLOID-pheatmap.png",
         plot   = bp3,
         res    = 600,
         width  = 6000,
         height = 6000)

setwd("..")

# pcaExplorer(dds = dds_ploidy)

beep(2)

```

# ALL treatments, triploids only
```{r ALL:TRT:diploids, warning=FALSE, include=TRUE}

# Filter data
coldata_trt <- coldata %>% filter(group == "B" | group == "D" | group == "F")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ group)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across 1/3 of the samples - https://support.bioconductor.org/p/110307/
keep <- rowSums(DESeq2::counts(dds) >= 10) >= ncol(cts_trt)/3
dds <- dds[keep,]

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)

bp0 <- biplot(p, showLoadings = FALSE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

bp0

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'group',
              colkey = c('A' = 'royalblue1' , 
                         'B' = 'green1'     ,
                         'C' = 'yellow2' ,
                         'D' = 'orange' ,
                         'E' = 'orangered1',
                         'F' = 'red3'),
              shape = 'ploidy',
              shapekey = c('diploid' = 19,  # circle
                           'triploid' = 17), # triangle
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              # ellipseFillKey = c('diploid' = 'blue', 'triploid' = 'green'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-140,140), ylim = c(-60, 70),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) + 
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1

bp2 <- pairsplot(p,
                components = getComponents(p, c(1:5)),
                triangle = TRUE, trianglelabSize = 12,
                hline = 0, vline = 0,
                pointSize = 1,
                gridlines.major = FALSE, gridlines.minor = FALSE,
                colby = 'group',
                title = 'Pairs plot', plotaxes = FALSE,
                margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2

# Plot pheatmap
rld <- rlog(dds, blind=T)
rld_mat <- assay(rld)
pca <- prcomp(t(rld_mat))
rld_cor <- cor(rld_mat)
bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])

bp3

setwd("DESEQ2_output/triploid")

ggexport(filename = "ALL-TREATMENTS-TRIPLOID-BIPLOT.png",
         plot   = bp0,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "ALL-TREATMENTS-TRIPLOID-PCA.png",
         plot   = bp1,
         res    = 600,
         # device = "png",
         width  = 4000,
         height = 5000)

ggexport(filename = "ALL-TREATMENTS-TRIPLOID-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

ggexport(filename = "ALL-TREATMENTS-TRIPLOID-pheatmap.png",
         plot   = bp3,
         res    = 600,
         width  = 6000,
         height = 6000)

# pcaExplorer(dds = dds_ploidy)
beep(2)
```

# CONTROLS only - effect of ploidy
```{r CONTROL:PLOIDY, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "A" | group == "B")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ ploidy)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across 1/3 of the samples - https://support.bioconductor.org/p/110307/
keep <- rowSums(DESeq2::counts(dds) >= 10) >= ncol(cts_trt)/3
dds <- dds[keep,]

# Generate Contrasts
contrast_list    <- c("ploidy", "triploid", "diploid")
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "DESEQ2_output/control_ploidy/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange results tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")

all_genes                  <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes                  <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment        <- "control_ploidy"

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "control_ploidy"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
all_genes_apeglm$treatment <- "control_ploidy"

all_genes_ashr             <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr             <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment   <- "control_ploidy"

# Filter results by lfc and pvalue cutoffs
sig_genes           <- res_table %>% filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes           <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes           <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "control_ploidy" 

sig_genes_normal           <- res_table_normal %>% filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal           <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal           <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "control_ploidy" 

sig_genes_apeglm <- res_table_apeglm %>% filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "control_ploidy" 

sig_genes_ashr <- res_table_ashr %>% filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "control_ploidy" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

# Write outputs
write.table(single_gene_expression_unshrunk, file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05_lfc1.5","DEG_normal-lfc1.5",
                            "DEG_apeglm-s0.005_lfc1.5","DEG_ashr-p0.05_lfc1.5")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "DESEQ2_output/control_ploidy/CONTROL_PLOIDY-gene-counts.csv",        row.names = TRUE)

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)

bp0 <- biplot(p, showLoadings = FALSE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

bp0

bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'ploidy',
              colkey = c('diploid' = 'royalblue1', 'triploid' = 'green1'),
              shape = 'ploidy',
              shapekey = c('diploid' = 19, 'triploid' = 17),
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              # ellipseFillKey = c('diploid' = 'blue', 'triploid' = 'green'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-125,145), ylim = c(-100, 80),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) +
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1

bp2 <- pairsplot(p,
                 components = getComponents(p, c(1:5)),
                 triangle = TRUE, trianglelabSize = 12,
                 hline = 0, vline = 0,
                 pointSize = 1,
                 gridlines.major = FALSE, gridlines.minor = FALSE,
                 colby = 'ploidy',
                 title = 'Pairs plot', plotaxes = FALSE,
                 margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2

# Plot pheatmap
rld <- rlog(dds, blind=T)
rld_mat <- assay(rld)
pca <- prcomp(t(rld_mat))
rld_cor <- cor(rld_mat)
bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])

bp3
 
setwd("DESEQ2_output/control_ploidy")

ggexport(filename = "control_ploidy-BIPLOT.png",
         plot   = bp0,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "control_ploidy-PCA.png",
         plot   = bp1,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "control_ploidy-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

ggexport(filename = "control_ploidy-pheatmap.png",
         plot   = bp3,
         res    = 600,
         width  = 6000,
         height = 6000)

setwd("..")
setwd("..")
 
# Gene expression heatmap
# Useful: https://www.datanovia.com/en/lessons/heatmap-in-r-static-and-interactive-visualization/#:~:text=%2C%20Colv%20%3D%20Colv)-,Complex%20heatmap,different%20data%20from%20different%20sources.
# From: https://github.com/mousepixels/sanbomics/blob/main/tutorial_complex_Heatmap.Rmd

# Grab all results from DESeq2 results table
all <- na.omit(res_table_apeglm)
sigs <- all %>% filter(padj < padj.cutoff)
# sigs <- all[all$baseMean > 50,]
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')
# filter by uncharacterized (remove for top-20 graph)
grx <- glob2rx("unchar*") 
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(grx, description)))
sigs_characterized   <- anti_join(sigs, sigs_uncharacterized, by = 'gene')

# Get list of genes and associated metadata
df <- as.data.frame(sigs_characterized)
row.names(df) <- df$gene
df_sigs <- as.data.frame(sigs)
row.names(df_sigs) <- df_sigs$gene
df_all <- as.data.frame(all)
row.names(df_all) <- df_all$gene
df.top <- df[ (df$baseMean > 10) & (abs(df$log2FoldChange) > 1.5),]
df.top <- df.top[order(df.top$log2FoldChange, decreasing = TRUE),]
# head(df.top)

# Get normalized count data from dds object
rlog_out <- rlog(dds, blind=FALSE) 
mat     <- assay(rlog_out)[rownames(df.top), rownames(coldata_trt)] #sig genes x samples
colnames(mat) <- rownames(coldata_trt)
base_mean <- rowMeans(mat)
mat_all <- assay(rlog_out)[rownames(df_all), rownames(coldata_trt)]
colnames(mat_all) <- rownames(coldata_trt)
mat_sigs <- assay(rlog_out)[rownames(df_sigs), rownames(coldata_trt)]
colnames(mat_sigs) <- rownames(coldata_trt)

# Generate Z-score
mat.scaled <- t(apply(mat, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled)<-colnames(mat)
mat.scaled_all <- t(apply(mat_all, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_all)<-colnames(mat_all)
mat.scaled_sigs <- t(apply(mat_sigs, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_sigs)<-colnames(mat_sigs)

# Keep the X number of genes
if(nrow(df.top)>20){num_keep <- 10} else {num_keep <- nrow(df.top)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(mat.scaled)-num_keep), nrow(mat.scaled)) )
l2_val <- as.matrix(df.top[rows_keep,]$log2FoldChange) #getting log2 value for each gene we are keeping
colnames(l2_val)<-"logFC"
mean <- as.matrix(df.top[rows_keep,]$baseMean) #getting mean value for each gene we are keeping
colnames(mean)<-"AveExpr"

#maps values between b/w/r for min and max l2 values
col_logFC <- colorRamp2(c(min(l2_val),0, max(l2_val)), c("blue", "white", "red"))

#maps between 0% quantile, and 75% quantile of mean values --- 0, 25, 50, 75, 100
col_AveExpr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white", "red"))

LOCID <- as.data.frame(rownames(mat.scaled[rows_keep,]))
colnames(LOCID) <- 'gene'
LOCID <- left_join(LOCID, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_complete <- as.data.frame(rownames(mat.scaled))
colnames(LOCID_complete) <- 'gene'
LOCID_complete <- left_join(LOCID_complete, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_sigs <- as.data.frame(rownames(mat.scaled_sigs))
colnames(LOCID_sigs) <- 'gene'
LOCID_sigs <- left_join(LOCID_sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

# Plot heatmap + annotations of top X number of DEGs (w/ lfc and pvalue cutoff)
png(filename = "DESEQ2_output/control_ploidy/control_ploidy_gene_heatmap.png", 
    res = 600, width = 6000, height = 4000, units = "px", pointsize = 6)
# Z-score matrix
h1 <- Heatmap(mat.scaled[rows_keep,], 
            column_labels       = colnames(mat.scaled),
            row_labels          = LOCID$description,
            row_names_side      = "left",
            row_names_gp        = gpar(fontsize = 8),
            row_names_centered  = F,
            cluster_columns     = F,
            cluster_rows        = F,
            row_names_max_width =  unit(10, "cm"))

# log2FoldChange matrix
h2 <- Heatmap(l2_val, row_labels = df.top$gene[rows_keep], 
            cluster_rows = F, 
            name="logFC", 
            col = col_logFC,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(l2_val[i, j],1), x, y)
            })

# Average expression
h3 <- Heatmap(mean, row_labels = df.top$symbol[rows_keep], 
            cluster_rows = F, 
            name = "AveExpr", 
            col=col_AveExpr,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(mean[i, j],0), x, y)
            })
h<-h1+h2+h3
h
# par(mar=c(5,8,4,1)+.1)
print(h)
dev.off()

# Plot heatmap
png(filename = "DESEQ2_output/control_ploidy/control_ploidy_heatmap_pval_genes_apeglm.png", 
    res = 1200, width = 6000, height = 10000, units = "px", pointsize = 6)
# par(mar=c(5,8,4,1)+.1)
col <- colorRampPalette(brewer.pal(11, "RdYlBu"))(256)
heatmap(mat.scaled_sigs, scale = "none", col =  col, 
        ColSideColors = c(rep("royalblue1", 11), rep("green1", 11)))
dev.off()

# Volcano plots
# All genes
p1 <- ggplot(all_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0), show.legend = FALSE) + my_theme

p2 <- ggplot(all_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0), show.legend = FALSE) + my_theme

p3 <- ggplot(all_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0), show.legend = FALSE) + my_theme

p4 <- ggplot(all_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0), show.legend = FALSE) + my_theme

# Significant Genes
p5 <- ggplot(sig_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0), show.legend = FALSE) + my_theme

p6 <- ggplot(sig_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0), show.legend = FALSE) + my_theme

p7 <- ggplot(sig_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0), show.legend = FALSE) + my_theme

p8 <- ggplot(sig_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0), show.legend = FALSE) + my_theme

# Save plots
ggsave("DESEQ2_output/control_ploidy/Volcano_all_genes.png",
       plot   = p1,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/control_ploidy/Volcano_all_genes_normal.png",
       plot   = p2,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/control_ploidy/Volcano_all_genes_ashr.png",
       plot   = p3,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/control_ploidy/Volcano_all_genes_apeglm.png",
       plot   = p4,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/control_ploidy/Volcano_pval_lfc_genes.png",
       plot   = p5,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/control_ploidy/Volcano_pval_lfc_genes_normal.png",
       plot   = p6,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/control_ploidy/Volcano_pval_lfc_genes_ashr.png",
       plot   = p7,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/control_ploidy/Volcano_pval_lfc_genes_apeglm.png",
       plot   = p8,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

### Pie chart of unknown vs known DEG
# Prepare dataframe for plot
all <- na.omit(res_table_apeglm) # Grab all DEGs, using apeglm shrinkage estimator
sigs <- all %>% filter(padj < padj.cutoff) # filter for significant DEGs
# sigs <- all[all$baseMean > 50,]
# sigs <- sigs[abs(sigs$log2FoldChange)>1.5,]
sigs$cat <- "characterized"
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene') # add information from gene master table
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(glob2rx("*unchar*"), description))) # find uncharacterized DEGs
if(nrow(sigs_uncharacterized) > 0){sigs_uncharacterized$cat <- "uncharacterized"} # add category label
sigs <- anti_join(sigs, sigs_uncharacterized,  by = "gene") %>% bind_rows(sigs_uncharacterized) # merge unchar and characterized

count_unchar <- as.numeric(nrow(sigs[sigs$cat == "uncharacterized",]))
count_char <- as.numeric(nrow(sigs[sigs$cat == "characterized",]))

sigs$count <- 1
bp4  <- ggplot(sigs, aes(x=factor(1),y=count, fill=cat)) +
        geom_bar(width = 1, stat = "identity", show.legend = FALSE) +
        scale_fill_manual(values = c("gray12",'azure4')) + labs(NULL) +
        coord_polar("y", start=0) + my_theme + theme(axis.title.x=element_blank(),
                                                     axis.title.y=element_blank(),
                                                     axis.text.x=element_blank(),
                                                     axis.text.y=element_blank(),
                                                     axis.ticks.y=element_blank())
bp4

### Pie chart of unknown of up or down regulated
sigs$plus_minus <- as.character(sign(sigs$log2FoldChange))
count_plus <- as.numeric(nrow(sigs[sigs$plus_minus == "1",]))
count_minus <- as.numeric(nrow(sigs[sigs$plus_minus == "-1",]))

bp5  <- ggplot(sigs, aes(x=factor(1),y=plus_minus, fill=plus_minus)) +
        geom_bar(width = 1, stat = "identity", show.legend = TRUE) +
        scale_fill_manual(values = c("lightskyblue","indianred")) + labs(NULL) +
        coord_polar("y", start=0) + my_theme
bp5

### Characterized sigs bar chart
sigs_pval_lfc <- sigs %>% filter(cat == "characterized") %>% filter(abs(log2FoldChange)>1.5) %>% 
  arrange(desc(log2FoldChange))
sigs_pval_lfc$gene <- factor(sigs_pval_lfc$gene, ordered = TRUE) 

# Keep the X number of genes (up/down)
num_keep = 10
if(nrow(sigs_pval_lfc)>20){num_keep <- 10} else {num_keep <- nrow(sigs_pval_lfc)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(sigs_pval_lfc)-num_keep+1), nrow(sigs_pval_lfc)))

bp6 <- ggplot(sigs_pval_lfc[rows_keep,], aes(x=as.numeric(row.names(sigs_pval_lfc[rows_keep,])), 
                                             y=log2FoldChange, fill=plus_minus)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             scale_fill_manual(values=c("salmon1","royalblue1")) +
             scale_y_continuous(breaks = seq(-4, 4, by = 2), limits = c(-5.5,5.5)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp6

bp7 <- ggplot(sigs_characterized, aes(x=as.numeric(row.names(sigs_characterized)), 
                                             y=log2FoldChange, fill=feature)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             # scale_fill_manual(values=c("salmon1","royalblue1")) +
             # scale_y_continuous(breaks = seq(-4, 8, by = 2), limits = c(-5.5,9)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp7

# volcano plot pval/lfc
p9 <- ggplot(sigs, aes(x = log2FoldChange, y = -log10(pvalue), color = plus_minus)) +
             geom_point(show.legend = FALSE) +
             scale_color_manual(values = c("salmon1","royalblue1")) +
             scale_x_continuous(breaks = seq(-4, 4, by = 2), limits = c(-5.5,5.5)) + 
             scale_y_continuous(breaks = seq(4, 20, by = 4), limits = c(2.5,20)) + my_theme + 
             theme(axis.title.x=element_blank(),axis.title.y=element_blank())

# Output tables of DEGs
sigs$analysis <- "control_ploidy"
sigs_pval_lfc$analysis <- "control_ploidy"
write.table(sigs, file = "DESEQ2_output/control_ploidy/control_ploidy-sig-genes-apeglm-pval-only.csv", row.names = F)
write.table(sigs_pval_lfc, file = "DESEQ2_output/control_ploidy/control_ploidy-sig-genes-apeglm-pval-lfc.csv", row.names = F)

# Save plots
ggsave("DESEQ2_output/control_ploidy/control_ploidy_DEG_uncharacterized_pval_apeglm.png",
       plot   = bp4, bg = "transparent",
       dpi    = 2500,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/control_ploidy/control_ploidy-DEG-apeglm-updown-pie.png",
       plot   = bp5,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/control_ploidy/control_ploidy-DEG-apeglm-updown-bar.png",
       plot   = bp6,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 5,
       units  = "in")

ggsave("DESEQ2_output/control_ploidy/control_ploidy-DEG-apeglm.png",
       plot   = bp7,
       dpi    = 600,
       device = "png",
       width  = 5,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/control_ploidy/Volcano_pval_genes_apeglm.png",
       plot   = p9,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")


beep(2)
```

# HEAT only - effect of ploidy
```{r HEAT:PLOIDY, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "C" | group == "D")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ ploidy)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across 1/3 of the samples - https://support.bioconductor.org/p/110307/
keep <- rowSums(DESeq2::counts(dds) >= 10) >= ncol(cts_trt)/3
dds <- dds[keep,]

# Generate Contrasts
contrast_list    <- c("ploidy", "triploid", "diploid")
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "DESEQ2_output/heat_ploidy/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "heat_ploidy"
# all_genes_pvalue    <- cbind(gene=all_genes$gene,padj=all_genes$padj)

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "heat_ploidy"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
# all_genes_apeglm           <- left_join(all_genes_apeglm, as.data.frame(all_genes_pvalue), by = "gene")
all_genes_apeglm$treatment <- "heat_ploidy"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "heat_ploidy"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "heat_ploidy" 

sig_genes_normal <- res_table_normal %>%
  filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "heat_ploidy" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "heat_ploidy" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "heat_ploidy" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05","DEG_normal-p0.05",
                            "DEG_apeglm-p0.05","DEG_ashr-p0.05")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "DESEQ2_output/heat_ploidy/HEAT_PLOIDY-gene-counts.csv",        row.names = TRUE)

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)

bp0 <- biplot(p, showLoadings = FALSE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

bp0

bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'group',
              colkey = c('A' = 'royalblue1' ,
                         'B' = 'green1'     ,
                         'C' = 'yellow2' ,
                         'D' = 'orange' ,
                         'E' = 'orangered1',
                         'F' = 'red3'),
              shape = 'ploidy',
              shapekey = c('diploid' = 19, 'triploid' = 17),
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              # ellipseFillKey = c('diploid' = 'blue', 'triploid' = 'green'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-100,110), ylim = c(-50, 60),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) +
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1

bp2 <- pairsplot(p,
                components = getComponents(p, c(1:5)),
                triangle = TRUE, trianglelabSize = 12,
                hline = 0, vline = 0,
                pointSize = 1,
                gridlines.major = FALSE, gridlines.minor = FALSE,
                colby = 'ploidy',
                title = 'Pairs plot', plotaxes = FALSE,
                margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2
 

# Plot pheatmap
rld <- rlog(dds, blind=T)
rld_mat <- assay(rld)
pca <- prcomp(t(rld_mat))
rld_cor <- cor(rld_mat)
bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])

bp3
 
 
setwd("DESEQ2_output/heat_ploidy")

ggexport(filename = "heat_ploidy-BIPLOT.png",
         plot   = bp0,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "heat_ploidy-PCA.png",
         plot   = bp1,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "heat_ploidy-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

ggexport(filename = "heat_ploidy-pheatmap.png",
         plot   = bp3,
         res    = 600,
         width  = 6000,
         height = 6000)

setwd("..")
setwd("..")
 
# Gene expression heatmap
# Useful: https://www.datanovia.com/en/lessons/heatmap-in-r-static-and-interactive-visualization/#:~:text=%2C%20Colv%20%3D%20Colv)-,Complex%20heatmap,different%20data%20from%20different%20sources.
# From: https://github.com/mousepixels/sanbomics/blob/main/tutorial_complex_Heatmap.Rmd

# Grab all results from DESeq2 results table
all <- na.omit(res_table_apeglm)
sigs <- all[all$padj < 0.05,]
# sigs <- all[all$baseMean > 50,]
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')
# filter by uncharacterized (remove for top-20 graph)
grx <- glob2rx("unchar*") 
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(grx, description)))
sigs_characterized   <- anti_join(sigs, sigs_uncharacterized, by = 'gene')

# Get list of genes and associated metadata
df <- as.data.frame(sigs_characterized)
row.names(df) <- df$gene
df_sigs <- as.data.frame(sigs)
row.names(df_sigs) <- df_sigs$gene
df_all <- as.data.frame(all)
row.names(df_all) <- df_all$gene
df.top <- df[ (df$baseMean > 10) & (abs(df$log2FoldChange) > 1.5),]
df.top <- df.top[order(df.top$log2FoldChange, decreasing = TRUE),]
# head(df.top)

# Get normalized count data from dds object
rlog_out <- rlog(dds, blind=FALSE) 
mat     <- assay(rlog_out)[rownames(df.top), rownames(coldata_trt)] #sig genes x samples
colnames(mat) <- rownames(coldata_trt)
base_mean <- rowMeans(mat)
mat_all <- assay(rlog_out)[rownames(df_all), rownames(coldata_trt)]
colnames(mat_all) <- rownames(coldata_trt)
mat_sigs <- assay(rlog_out)[rownames(df_sigs), rownames(coldata_trt)]
colnames(mat_sigs) <- rownames(coldata_trt)

# Generate Z-score
mat.scaled <- t(apply(mat, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled)<-colnames(mat)
mat.scaled_all <- t(apply(mat_all, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_all)<-colnames(mat_all)
mat.scaled_sigs <- t(apply(mat_sigs, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_sigs)<-colnames(mat_sigs)

# Keep the X number of  
if(nrow(df.top)>20){num_keep <- 10} else {num_keep <- nrow(df.top)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(mat.scaled)-num_keep), nrow(mat.scaled)) )
l2_val <- as.matrix(df.top[rows_keep,]$log2FoldChange) #getting log2 value for each gene we are keeping
colnames(l2_val)<-"logFC"
mean <- as.matrix(df.top[rows_keep,]$baseMean) #getting mean value for each gene we are keeping
colnames(mean)<-"AveExpr"

#maps values between b/w/r for min and max l2 values
col_logFC <- colorRamp2(c(min(l2_val),0, max(l2_val)), c("blue", "white", "red"))

#maps between 0% quantile, and 75% quantile of mean values --- 0, 25, 50, 75, 100
col_AveExpr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white", "red"))

LOCID <- as.data.frame(rownames(mat.scaled[rows_keep,]))
colnames(LOCID) <- 'gene'
LOCID <- left_join(LOCID, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_complete <- as.data.frame(rownames(mat.scaled))
colnames(LOCID_complete) <- 'gene'
LOCID_complete <- left_join(LOCID_complete, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_sigs <- as.data.frame(rownames(mat.scaled_sigs))
colnames(LOCID_sigs) <- 'gene'
LOCID_sigs <- left_join(LOCID_sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

# Plot heatmap + annotations of top X number of DEGs (w/ lfc and pvalue cutoff)
png(filename = "DESEQ2_output/heat_ploidy/heat_ploidy_gene_heatmap.png", 
    res = 600, width = 6000, height = 4000, units = "px", pointsize = 6)
# Z-score matrix
h1 <- Heatmap(mat.scaled[rows_keep,], 
            column_labels       = colnames(mat.scaled),
            row_labels          = LOCID$description,
            row_names_side      = "left",
            row_names_gp        = gpar(fontsize = 8),
            row_names_centered  = F,
            cluster_columns     = F,
            cluster_rows        = F,
            row_names_max_width =  unit(10, "cm"))

# log2FoldChange matrix
h2 <- Heatmap(l2_val, row_labels = df.top$gene[rows_keep], 
            cluster_rows = F, 
            name="logFC", 
            col = col_logFC,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(l2_val[i, j],1), x, y)
            })

# Average expression
h3 <- Heatmap(mean, row_labels = df.top$symbol[rows_keep], 
            cluster_rows = F, 
            name = "AveExpr", 
            col=col_AveExpr,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(mean[i, j],0), x, y)
            })
h<-h1+h2+h3
h
# par(mar=c(5,8,4,1)+.1)
print(h)
dev.off()

# Plot heatmap
png(filename = "DESEQ2_output/heat_ploidy/heat_ploidy_heatmap_pval_genes_apeglm.png", 
    res = 1200, width = 6000, height = 6000, units = "px", pointsize = 6)
# par(mar=c(5,8,4,1)+.1)
col <- colorRampPalette(brewer.pal(11, "RdYlBu"))(256)
heatmap(mat.scaled_sigs, scale = "none", col =  col, 
        ColSideColors = c(rep("royalblue1", 11), rep("green1", 10)))
dev.off()

# Volcano plots
# All genes
p1 <- ggplot(all_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p2 <- ggplot(all_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p3 <- ggplot(all_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p4 <- ggplot(all_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

# Significant Genes
p5 <- ggplot(sig_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p6 <- ggplot(sig_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p7 <- ggplot(sig_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p8 <- ggplot(sig_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

# Save plots
ggsave("DESEQ2_output/heat_ploidy/Volcano_all_genes.png",
       plot   = p1,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/heat_ploidy/Volcano_all_genes_normal.png",
       plot   = p2,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/heat_ploidy/Volcano_all_genes_ashr.png",
       plot   = p3,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/heat_ploidy/Volcano_all_genes_apeglm.png",
       plot   = p4,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/heat_ploidy/Volcano_pval_lfc_genes.png",
       plot   = p5,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/heat_ploidy/Volcano_pval_lfc_genes_normal.png",
       plot   = p6,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/heat_ploidy/Volcano_pval_lfc_genes_ashr.png",
       plot   = p7,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/heat_ploidy/Volcano_pval_lfc_genes_apeglm.png",
       plot   = p8,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

### Pie chart of unknown vs known DEG
# Prepare dataframe for plot
all <- na.omit(res_table_apeglm) # Grab all DEGs, using apeglm shrinkage estimator
sigs <- all[all$padj < 0.05,] # filter for significant DEGs
# sigs <- all[all$baseMean > 50,]
# sigs <- sigs[abs(sigs$log2FoldChange)>1.5,]
sigs$cat <- "characterized"
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene') # add information from gene master table
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(glob2rx("*unchar*"), description))) # find uncharacterized DEGs
if(nrow(sigs_uncharacterized) > 0){sigs_uncharacterized$cat <- "uncharacterized"} # add category label
sigs <- anti_join(sigs, sigs_uncharacterized,  by = "gene") %>% bind_rows(sigs_uncharacterized) # merge unchar and characterized

sigs$count <- 1
bp4  <- ggplot(sigs, aes(x=factor(1),y=count, fill=cat)) +
        geom_bar(width = 1, stat = "identity", show.legend = FALSE) +
        scale_fill_manual(values = c("gray12",'azure4')) + labs(NULL) +
        coord_polar("y", start=0) + my_theme + theme(axis.title.x=element_blank(),
                                                     axis.title.y=element_blank(),
                                                     axis.text.x=element_blank(),
                                                     axis.text.y=element_blank(),
                                                     axis.ticks.y=element_blank())
bp4

### Pie chart of unknown of up or down regulated
sigs$plus_minus <- as.character(sign(sigs$log2FoldChange))

bp5  <- ggplot(sigs, aes(x=factor(1),y=plus_minus, fill=plus_minus)) +
        geom_bar(width = 1, stat = "identity", show.legend = FALSE) +
        scale_fill_manual(values = c("lightskyblue","indianred")) + labs(NULL) +
        coord_polar("y", start=0) + my_theme
bp5

### Characterized sigs bar chart
sigs_pval_lfc <- sigs %>% filter(cat == "characterized") %>% filter(abs(log2FoldChange)>1.5) %>% 
  arrange(desc(log2FoldChange))
sigs_pval_lfc$gene <- factor(sigs_pval_lfc$gene, ordered = TRUE) 

# Keep the X number of genes (up/down)
num_keep = 10
if(nrow(sigs_pval_lfc)>20){num_keep <- 10} else {num_keep <- nrow(sigs_pval_lfc)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(sigs_pval_lfc)-num_keep+1), nrow(sigs_pval_lfc)))

bp6 <- ggplot(sigs_pval_lfc[rows_keep,], aes(x=as.numeric(row.names(sigs_pval_lfc[rows_keep,])), 
                                             y=log2FoldChange, fill=plus_minus)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             scale_fill_manual(values=c("salmon1","royalblue1")) +
             scale_y_continuous(breaks = seq(-5, 5, by = 2), limits = c(-5,5)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp6

bp7 <- ggplot(sigs_characterized, aes(x=as.numeric(row.names(sigs_characterized)), 
                                             y=log2FoldChange, fill=feature)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             # scale_fill_manual(values=c("salmon1","royalblue1")) +
             # scale_y_continuous(breaks = seq(-4, 8, by = 2), limits = c(-5.5,9)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp7

# volcano plot pval/lfc
p9 <- ggplot(sigs, aes(x = log2FoldChange, y = -log10(pvalue), color = plus_minus)) +
             geom_point(show.legend = FALSE) +
             scale_color_manual(values = c("salmon1","royalblue1")) +
             scale_x_continuous(breaks = seq(-4, 4, by = 2), limits = c(-5.5,5.5)) + 
             scale_y_continuous(breaks = seq(4, 20, by = 4), limits = c(2.5,20)) + my_theme + 
             theme(axis.title.x=element_blank(),axis.title.y=element_blank())

# Output tables of DEGs
sigs$analysis <- "heat_ploidy"
sigs_pval_lfc$analysis <- "heat_ploidy"
write.table(sigs, file = "DESEQ2_output/heat_ploidy/heat_ploidy-sig-genes-apeglm-pval-only.csv", row.names = F)
write.table(sigs_pval_lfc, file = "DESEQ2_output/heat_ploidy/heat_ploidy-sig-genes-apeglm-pval-lfc.csv", row.names = F)

# Save plots
ggsave("DESEQ2_output/heat_ploidy/heat_ploidy-DEG-uncharacterized-pval-apeglm.png",
       plot   = bp4, bg = "transparent",
       dpi    = 1200,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/heat_ploidy/heat_ploidy-DEG-apeglm-updown-pie.png",
       plot   = bp5,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/heat_ploidy/heat_ploidy-DEG-apeglm-updown-bar.png",
       plot   = bp6,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/heat_ploidy/heat_ploidy-DEG-apeglm.png",
       plot   = bp7,
       dpi    = 600,
       device = "png",
       width  = 5,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/heat_ploidy/heat_ploidy-volcano_pval_genes_apeglm.png",
       plot   = p9,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

beep(2)
```

# DESICCATION only - effect of ploidy
```{r DESICCATION:PLOIDY, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "E" | group == "F")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ ploidy)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across 1/3 of the samples - https://support.bioconductor.org/p/110307/
keep <- rowSums(DESeq2::counts(dds) >= 10) >= ncol(cts_trt)/3
dds <- dds[keep,]

# Generate Contrasts
contrast_list    <- c("ploidy", "triploid", "diploid")
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "DESEQ2_output/desiccation_ploidy/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "desiccation_ploidy"
# all_genes_pvalue    <- cbind(gene=all_genes$gene,padj=all_genes$padj)

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "desiccation_ploidy"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
# all_genes_apeglm           <- left_join(all_genes_apeglm, as.data.frame(all_genes_pvalue), by = "gene")
all_genes_apeglm$treatment <- "desiccation_ploidy"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "desiccation_ploidy"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "desiccation_ploidy" 

sig_genes_normal <- res_table_normal %>%
  filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "desiccation_ploidy" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "desiccation_ploidy" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "desiccation_ploidy" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05","DEG_normal-p0.05",
                            "DEG_apeglm-p0.05","DEG_ashr-p0.05")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-gene-counts.csv",        row.names = TRUE)

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)

bp0 <- biplot(p, showLoadings = FALSE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

bp0

bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'group',
              colkey = c('A' = 'royalblue1' ,
                         'B' = 'green1'     ,
                         'C' = 'yellow2' ,
                         'D' = 'orange' ,
                         'E' = 'orangered1',
                         'F' = 'red3'),
              shape = 'ploidy',
              shapekey = c('diploid' = 19, 'triploid' = 17),
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              # ellipseFillKey = c('diploid' = 'blue', 'triploid' = 'green'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-150,120), ylim = c(-40, 50),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) +
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1

bp2 <- pairsplot(p,
                components = getComponents(p, c(1:5)),
                triangle = TRUE, trianglelabSize = 12,
                hline = 0, vline = 0,
                pointSize = 1,
                gridlines.major = FALSE, gridlines.minor = FALSE,
                colby = 'ploidy',
                title = 'Pairs plot', plotaxes = FALSE,
                margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2
 
# Plot pheatmap
rld <- rlog(dds, blind=T)
rld_mat <- assay(rld)
pca <- prcomp(t(rld_mat))
rld_cor <- cor(rld_mat)
bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])

bp3
 
 
setwd("DESEQ2_output/desiccation_ploidy")

ggexport(filename = "desiccation_ploidy-BIPLOT.png",
         plot   = bp0,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "desiccation_ploidy-PCA.png",
         plot   = bp1,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "desiccation_ploidy-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

ggexport(filename = "desiccation_ploidy-pheatmap.png",
         plot   = bp3,
         res    = 600,
         width  = 6000,
         height = 6000)

setwd("..")
setwd("..")
 
# Gene expression heatmap
# Useful: https://www.datanovia.com/en/lessons/heatmap-in-r-static-and-interactive-visualization/#:~:text=%2C%20Colv%20%3D%20Colv)-,Complex%20heatmap,different%20data%20from%20different%20sources.
# From: https://github.com/mousepixels/sanbomics/blob/main/tutorial_complex_Heatmap.Rmd

# Grab all results from DESeq2 results table
all <- na.omit(res_table_apeglm)
sigs <- all[all$padj < 0.05,]
# sigs <- all[all$baseMean > 50,]
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')
# filter by uncharacterized (remove for top-20 graph)
grx <- glob2rx("unchar*") 
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(grx, description)))
sigs_characterized   <- anti_join(sigs, sigs_uncharacterized, by = 'gene')

# Get list of genes and associated metadata
df <- as.data.frame(sigs_characterized)
row.names(df) <- df$gene
df_sigs <- as.data.frame(sigs)
row.names(df_sigs) <- df_sigs$gene
df_all <- as.data.frame(all)
row.names(df_all) <- df_all$gene
df.top <- df[ (df$baseMean > 10) & (abs(df$log2FoldChange) > 1.5),]
df.top <- df.top[order(df.top$log2FoldChange, decreasing = TRUE),]
# head(df.top)

# Get normalized count data from dds object
rlog_out <- rlog(dds, blind=FALSE) 
mat     <- assay(rlog_out)[rownames(df.top), rownames(coldata_trt)] #sig genes x samples
colnames(mat) <- rownames(coldata_trt)
base_mean <- rowMeans(mat)
mat_all <- assay(rlog_out)[rownames(df_all), rownames(coldata_trt)]
colnames(mat_all) <- rownames(coldata_trt)
mat_sigs <- assay(rlog_out)[rownames(df_sigs), rownames(coldata_trt)]
colnames(mat_sigs) <- rownames(coldata_trt)

# Generate Z-score
mat.scaled <- t(apply(mat, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled)<-colnames(mat)
mat.scaled_all <- t(apply(mat_all, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_all)<-colnames(mat_all)
mat.scaled_sigs <- t(apply(mat_sigs, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_sigs)<-colnames(mat_sigs)

# Keep the X number of  
if(nrow(df.top)>20){num_keep <- 10} else {num_keep <- nrow(df.top)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(mat.scaled)-num_keep), nrow(mat.scaled)) )
l2_val <- as.matrix(df.top[rows_keep,]$log2FoldChange) #getting log2 value for each gene we are keeping
colnames(l2_val)<-"logFC"
mean <- as.matrix(df.top[rows_keep,]$baseMean) #getting mean value for each gene we are keeping
colnames(mean)<-"AveExpr"

#maps values between b/w/r for min and max l2 values
col_logFC <- colorRamp2(c(min(l2_val),0, max(l2_val)), c("blue", "white", "red"))

#maps between 0% quantile, and 75% quantile of mean values --- 0, 25, 50, 75, 100
col_AveExpr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white", "red"))

LOCID <- as.data.frame(rownames(mat.scaled[rows_keep,]))
colnames(LOCID) <- 'gene'
LOCID <- left_join(LOCID, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_complete <- as.data.frame(rownames(mat.scaled))
colnames(LOCID_complete) <- 'gene'
LOCID_complete <- left_join(LOCID_complete, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_sigs <- as.data.frame(rownames(mat.scaled_sigs))
colnames(LOCID_sigs) <- 'gene'
LOCID_sigs <- left_join(LOCID_sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

# Plot heatmap + annotations of top X number of DEGs (w/ lfc and pvalue cutoff)
png(filename = "DESEQ2_output/desiccation_ploidy/desiccation_ploidy_gene_heatmap.png", 
    res = 600, width = 6000, height = 4000, units = "px", pointsize = 6)
# Z-score matrix
h1 <- Heatmap(mat.scaled[rows_keep,], 
            column_labels       = colnames(mat.scaled),
            row_labels          = LOCID$description,
            row_names_side      = "left",
            row_names_gp        = gpar(fontsize = 8),
            row_names_centered  = F,
            cluster_columns     = F,
            cluster_rows        = F,
            row_names_max_width =  unit(10, "cm"))

# log2FoldChange matrix
h2 <- Heatmap(l2_val, row_labels = df.top$gene[rows_keep], 
            cluster_rows = F, 
            name="logFC", 
            col = col_logFC,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(l2_val[i, j],1), x, y)
            })

# Average expression
h3 <- Heatmap(mean, row_labels = df.top$symbol[rows_keep], 
            cluster_rows = F, 
            name = "AveExpr", 
            col=col_AveExpr,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(mean[i, j],0), x, y)
            })
h<-h1+h2+h3
h
# par(mar=c(5,8,4,1)+.1)
print(h)
dev.off()

# Plot heatmap
png(filename = "DESEQ2_output/desiccation_ploidy/desiccation_ploidy_heatmap_pval_genes_apeglm.png", 
    res = 1200, width = 6000, height = 6000, units = "px", pointsize = 6)
# par(mar=c(5,8,4,1)+.1)
col <- colorRampPalette(brewer.pal(11, "RdYlBu"))(256)
heatmap(mat.scaled_sigs, scale = "none", col =  col, 
        ColSideColors = c(rep("royalblue1", 10), rep("green1", 11)))
dev.off()

# Volcano plots
# All genes
p1 <- ggplot(all_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p2 <- ggplot(all_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p3 <- ggplot(all_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p4 <- ggplot(all_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

# Significant Genes
p5 <- ggplot(sig_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p6 <- ggplot(sig_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p7 <- ggplot(sig_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p8 <- ggplot(sig_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

# Save plots
ggsave("DESEQ2_output/desiccation_ploidy/Volcano_all_genes.png",
       plot   = p1,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/desiccation_ploidy/Volcano_all_genes_normal.png",
       plot   = p2,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/desiccation_ploidy/Volcano_all_genes_ashr.png",
       plot   = p3,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/desiccation_ploidy/Volcano_all_genes_apeglm.png",
       plot   = p4,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/desiccation_ploidy/Volcano_pval_lfc_genes.png",
       plot   = p5,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/desiccation_ploidy/Volcano_pval_lfc_genes_normal.png",
       plot   = p6,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/desiccation_ploidy/Volcano_pval_lfc_genes_ashr.png",
       plot   = p7,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/desiccation_ploidy/Volcano_pval_lfc_genes_apeglm.png",
       plot   = p8,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

### Pie chart of unknown vs known DEG
# Prepare dataframe for plot
all <- na.omit(res_table_apeglm) # Grab all DEGs, using apeglm shrinkage estimator
sigs <- all[all$padj < 0.05,] # filter for significant DEGs
# sigs <- all[all$baseMean > 50,]
# sigs <- sigs[abs(sigs$log2FoldChange)>1.5,]
sigs$cat <- "characterized"
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene') # add information from gene master table
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(glob2rx("*unchar*"), description))) # find uncharacterized DEGs
if(nrow(sigs_uncharacterized) > 0){sigs_uncharacterized$cat <- "uncharacterized"} # add category label
sigs <- anti_join(sigs, sigs_uncharacterized,  by = "gene") %>% bind_rows(sigs_uncharacterized) # merge unchar and characterized

count_unchar <- as.numeric(nrow(sigs[sigs$cat == "uncharacterized",]))
count_char <- as.numeric(nrow(sigs[sigs$cat == "characterized",]))

sigs$count <- 1
bp4  <- ggplot(sigs, aes(x=factor(1),y=count, fill=cat)) +
        geom_bar(width = 1, stat = "identity", show.legend = FALSE) +
        scale_fill_manual(values = c("gray12",'azure4')) + labs(NULL) +
        coord_polar("y", start=0) + my_theme + theme(axis.title.x=element_blank(),
                                                     axis.title.y=element_blank(),
                                                     axis.text.x=element_blank(),
                                                     axis.text.y=element_blank(),
                                                     axis.ticks.y=element_blank())
bp4

### Pie chart of unknown of up or down regulated
sigs$plus_minus <- as.character(sign(sigs$log2FoldChange))
count_plus <- as.numeric(nrow(sigs[sigs$plus_minus == "1",]))
count_minus <- as.numeric(nrow(sigs[sigs$plus_minus == "-1",]))

bp5  <- ggplot(sigs, aes(x=factor(1),y=plus_minus, fill=plus_minus)) +
        geom_bar(width = 1, stat = "identity", show.legend = TRUE) +
        scale_fill_manual(values = c("lightskyblue","indianred")) + labs(NULL) +
        coord_polar("y", start=0) + my_theme
bp5

### Characterized sigs bar chart
sigs_pval_lfc <- sigs %>% filter(cat == "characterized") %>% filter(abs(log2FoldChange)>1.5) %>% 
  arrange(desc(log2FoldChange))
sigs_pval_lfc$gene <- factor(sigs_pval_lfc$gene, ordered = TRUE) 

# Keep the X number of genes (up/down)
num_keep = 10
if(nrow(sigs_pval_lfc)>20){num_keep <- 10} else {num_keep <- nrow(sigs_pval_lfc)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(sigs_pval_lfc)-num_keep+1), nrow(sigs_pval_lfc)))

bp6 <- ggplot(sigs_pval_lfc[rows_keep,], aes(x=as.numeric(row.names(sigs_pval_lfc[rows_keep,])), 
                                             y=log2FoldChange, fill=plus_minus)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             scale_fill_manual(values=c("salmon1","royalblue1")) +
             scale_y_continuous(breaks = seq(-4, 4, by = 2), limits = c(-5.5,5.5)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp6

bp7 <- ggplot(sigs_characterized, aes(x=as.numeric(row.names(sigs_characterized)), 
                                             y=log2FoldChange, fill=feature)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             # scale_fill_manual(values=c("salmon1","royalblue1")) +
             # scale_y_continuous(breaks = seq(-4, 8, by = 2), limits = c(-5.5,9)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp7

# volcano plot pval/lfc
p9 <- ggplot(sigs, aes(x = log2FoldChange, y = -log10(pvalue), color = plus_minus)) +
             geom_point(show.legend = FALSE) +
             scale_color_manual(values = c("salmon1","royalblue1")) +
             scale_x_continuous(breaks = seq(-4, 4, by = 2), limits = c(-5.5,5.5)) + 
             scale_y_continuous(breaks = seq(4, 20, by = 4), limits = c(2.5,20)) + my_theme + 
             theme(axis.title.x=element_blank(),axis.title.y=element_blank())

# Output tables of DEGs
sigs$analysis <- "desiccation_ploidy"
sigs_pval_lfc$analysis <- "desiccation_ploidy"
write.table(sigs, file = "DESEQ2_output/desiccation_ploidy/desiccation_ploidy-sig-genes-apeglm-pval-only.csv", row.names = F)
write.table(sigs_pval_lfc, file = "DESEQ2_output/desiccation_ploidy/desiccation_ploidy-sig-genes-apeglm-pval-lfc.csv", row.names = F)

# Save plots
ggsave("DESEQ2_output/desiccation_ploidy/desiccation_ploidy_DEG_uncharacterized_pval_apeglm.png",
       plot   = bp4, bg = "transparent",
       dpi    = 2500,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/desiccation_ploidy/desiccation_ploidy-DEG-apeglm-updown-pie.png",
       plot   = bp5,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/desiccation_ploidy/desiccation_ploidy-DEG-apeglm-updown-bar.png",
       plot   = bp6,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 5,
       units  = "in")

ggsave("DESEQ2_output/desiccation_ploidy/desiccation_ploidy-DEG-apeglm.png",
       plot   = bp7,
       dpi    = 600,
       device = "png",
       width  = 5,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/desiccation_ploidy/desiccation_ploidy_Volcano_pval_genes_apeglm.png",
       plot   = p9,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

beep(2)
```

# DIPLOID only - effect of temperature
```{r DIPLOID:HEAT, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "A" | group == "C")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ condition)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across 1/3 of the samples - https://support.bioconductor.org/p/110307/
keep <- rowSums(DESeq2::counts(dds) >= 10) >= ncol(cts_trt)/3
dds <- dds[keep,]

# Generate Contrasts
contrast_list    <- c("condition", "heat", "control") # factor, treatment group, control
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "DESEQ2_output/diploid_heat/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "diploid_heat"
# all_genes_pvalue    <- cbind(gene=all_genes$gene,padj=all_genes$padj)

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "diploid_heat"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
# all_genes_apeglm           <- left_join(all_genes_apeglm, as.data.frame(all_genes_pvalue), by = "gene")
all_genes_apeglm$treatment <- "diploid_heat"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "diploid_heat"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "diploid_heat" 

sig_genes_normal <- res_table_normal %>%
  filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "diploid_heat" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "diploid_heat" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "diploid_heat" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05","DEG_normal-p0.05",
                            "DEG_apeglm-p0.05","DEG_ashr-p0.05")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "DESEQ2_output/diploid_heat/DIPLOID_HEAT-gene-counts.csv",        row.names = TRUE)

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)

bp0 <- biplot(p, showLoadings = FALSE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

bp0

bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'group',
              colkey = c('A' = 'royalblue1' ,
                         'B' = 'green1'     ,
                         'C' = 'yellow2' ,
                         'D' = 'orange' ,
                         'E' = 'orangered1',
                         'F' = 'red3'),
              shape = 'condition',
              shapekey = c('control' = 19, 'heat' = 19),
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              ellipseFillKey = c('control' = 'royalblue1', 'heat' = 'orangered1'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-140,100), ylim = c(-70, 50),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) +
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1
# 
bp2 <- pairsplot(p,
                components = getComponents(p, c(1:5)),
                triangle = TRUE, trianglelabSize = 12,
                hline = 0, vline = 0,
                pointSize = 1,
                gridlines.major = FALSE, gridlines.minor = FALSE,
                colby = 'condition',
                title = 'Pairs plot', plotaxes = FALSE,
                margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2

# Plot pheatmap
rld <- rlog(dds, blind=T)
rld_mat <- assay(rld)
pca <- prcomp(t(rld_mat))
rld_cor <- cor(rld_mat)
bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])

bp3
 
 
setwd("DESEQ2_output/diploid_heat")

ggexport(filename = "diploid_heat-BIPLOT.png",
         plot   = bp0,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "diploid_heat-PCA.png",
         plot   = bp1,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "diploid_heat-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

ggexport(filename = "diploid_heat-pheatmap.png",
         plot   = bp3,
         res    = 600,
         width  = 6000,
         height = 6000)

setwd("..")
setwd("..")
 
# Gene expression heatmap
# Useful: https://www.datanovia.com/en/lessons/heatmap-in-r-static-and-interactive-visualization/#:~:text=%2C%20Colv%20%3D%20Colv)-,Complex%20heatmap,different%20data%20from%20different%20sources.
# From: https://github.com/mousepixels/sanbomics/blob/main/tutorial_complex_Heatmap.Rmd

# Grab all results from DESeq2 results table
all <- na.omit(res_table_apeglm)
sigs <- all[all$padj < 0.05,]
# sigs <- all[all$baseMean > 50,]
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')
# filter by uncharacterized (remove for top-20 graph)
grx <- glob2rx("unchar*") 
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(grx, description)))
sigs_characterized   <- anti_join(sigs, sigs_uncharacterized, by = 'gene')

# Get list of genes and associated metadata
df <- as.data.frame(sigs_characterized)
row.names(df) <- df$gene
df_sigs <- as.data.frame(sigs)
row.names(df_sigs) <- df_sigs$gene
df_all <- as.data.frame(all)
row.names(df_all) <- df_all$gene
df.top <- df[ (df$baseMean > 10) & (abs(df$log2FoldChange) > 1.5),]
df.top <- df.top[order(df.top$log2FoldChange, decreasing = TRUE),]
# head(df.top)

# Get normalized count data from dds object
rlog_out <- rlog(dds, blind=FALSE) 
mat     <- assay(rlog_out)[rownames(df.top), rownames(coldata_trt)] #sig genes x samples
colnames(mat) <- rownames(coldata_trt)
base_mean <- rowMeans(mat)
mat_all <- assay(rlog_out)[rownames(df_all), rownames(coldata_trt)]
colnames(mat_all) <- rownames(coldata_trt)
mat_sigs <- assay(rlog_out)[rownames(df_sigs), rownames(coldata_trt)]
colnames(mat_sigs) <- rownames(coldata_trt)

# Generate Z-score
mat.scaled <- t(apply(mat, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled)<-colnames(mat)
mat.scaled_all <- t(apply(mat_all, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_all)<-colnames(mat_all)
mat.scaled_sigs <- t(apply(mat_sigs, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_sigs)<-colnames(mat_sigs)

# Keep the X number of  
if(nrow(df.top)>20){num_keep <- 10} else {num_keep <- nrow(df.top)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(mat.scaled)-num_keep), nrow(mat.scaled)) )
l2_val <- as.matrix(df.top[rows_keep,]$log2FoldChange) #getting log2 value for each gene we are keeping
colnames(l2_val)<-"logFC"
mean <- as.matrix(df.top[rows_keep,]$baseMean) #getting mean value for each gene we are keeping
colnames(mean)<-"AveExpr"

#maps values between b/w/r for min and max l2 values
col_logFC <- colorRamp2(c(min(l2_val),0, max(l2_val)), c("blue", "white", "red"))

#maps between 0% quantile, and 75% quantile of mean values --- 0, 25, 50, 75, 100
col_AveExpr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white", "red"))

LOCID <- as.data.frame(rownames(mat.scaled[rows_keep,]))
colnames(LOCID) <- 'gene'
LOCID <- left_join(LOCID, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_complete <- as.data.frame(rownames(mat.scaled))
colnames(LOCID_complete) <- 'gene'
LOCID_complete <- left_join(LOCID_complete, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_sigs <- as.data.frame(rownames(mat.scaled_sigs))
colnames(LOCID_sigs) <- 'gene'
LOCID_sigs <- left_join(LOCID_sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

# Plot heatmap + annotations of top X number of DEGs (w/ lfc and pvalue cutoff)
png(filename = "DESEQ2_output/diploid_heat/diploid_heat_gene_heatmap.png", 
    res = 600, width = 6000, height = 4000, units = "px", pointsize = 6)
# Z-score matrix
h1 <- Heatmap(mat.scaled[rows_keep,], 
            column_labels       = colnames(mat.scaled),
            row_labels          = LOCID$description,
            row_names_side      = "left",
            row_names_gp        = gpar(fontsize = 8),
            row_names_centered  = F,
            cluster_columns     = F,
            cluster_rows        = F,
            row_names_max_width =  unit(10, "cm"))

# log2FoldChange matrix
h2 <- Heatmap(l2_val, row_labels = df.top$gene[rows_keep], 
            cluster_rows = F, 
            name="logFC", 
            col = col_logFC,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(l2_val[i, j],1), x, y)
            })

# Average expression
h3 <- Heatmap(mean, row_labels = df.top$symbol[rows_keep], 
            cluster_rows = F, 
            name = "AveExpr", 
            col=col_AveExpr,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(mean[i, j],0), x, y)
            })
h<-h1+h2+h3
h
# par(mar=c(5,8,4,1)+.1)
print(h)
dev.off()

# Plot heatmap
png(filename = "DESEQ2_output/diploid_heat/diploid_heat_heatmap_pval_genes_apeglm.png", 
    res = 1200, width = 6000, height = 6000, units = "px", pointsize = 6)
# par(mar=c(5,8,4,1)+.1)
col <- colorRampPalette(brewer.pal(11, "RdYlBu"))(256)
heatmap(mat.scaled_sigs, scale = "none", col =  col, 
        ColSideColors = c(rep("royalblue1", 11), rep("orangered1", 11)))
dev.off()

# Volcano plots
# All genes
p1 <- ggplot(all_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p2 <- ggplot(all_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p3 <- ggplot(all_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p4 <- ggplot(all_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

# Significant Genes
p5 <- ggplot(sig_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p6 <- ggplot(sig_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p7 <- ggplot(sig_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p8 <- ggplot(sig_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

# Save plots
ggsave("DESEQ2_output/diploid_heat/Volcano_all_genes.png",
       plot   = p1,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_heat/Volcano_all_genes_normal.png",
       plot   = p2,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_heat/Volcano_all_genes_ashr.png",
       plot   = p3,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_heat/Volcano_all_genes_apeglm.png",
       plot   = p4,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_heat/Volcano_pval_lfc_genes.png",
       plot   = p5,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_heat/Volcano_pval_lfc_genes_normal.png",
       plot   = p6,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_heat/Volcano_pval_lfc_genes_ashr.png",
       plot   = p7,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_heat/Volcano_pval_lfc_genes_apeglm.png",
       plot   = p8,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

### Pie chart of unknown vs known DEG
# Prepare dataframe for plot
all <- na.omit(res_table_apeglm) # Grab all DEGs, using apeglm shrinkage estimator
sigs <- all[all$padj < 0.05,] # filter for significant DEGs
# sigs <- all[all$baseMean > 50,]
# sigs <- sigs[abs(sigs$log2FoldChange)>1.5,]
sigs$cat <- "characterized"
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene') # add information from gene master table
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(glob2rx("*unchar*"), description))) # find uncharacterized DEGs
if(nrow(sigs_uncharacterized) > 0){sigs_uncharacterized$cat <- "uncharacterized"} # add category label
sigs <- anti_join(sigs, sigs_uncharacterized,  by = "gene") %>% bind_rows(sigs_uncharacterized) # merge unchar and characterized

count_unchar <- as.numeric(nrow(sigs[sigs$cat == "uncharacterized",]))
count_char <- as.numeric(nrow(sigs[sigs$cat == "characterized",]))

sigs$count <- 1
bp4  <- ggplot(sigs, aes(x=factor(1),y=count, fill=cat)) +
        geom_bar(width = 1, stat = "identity", show.legend = FALSE) +
        scale_fill_manual(values = c("gray12",'azure4')) + labs(NULL) +
        coord_polar("y", start=0) + my_theme + theme(axis.title.x=element_blank(),
                                                     axis.title.y=element_blank(),
                                                     axis.text.x=element_blank(),
                                                     axis.text.y=element_blank(),
                                                     axis.ticks.y=element_blank())
bp4

### Pie chart of unknown of up or down regulated
sigs$plus_minus <- as.character(sign(sigs$log2FoldChange))
count_plus <- as.numeric(nrow(sigs[sigs$plus_minus == "1",]))
count_minus <- as.numeric(nrow(sigs[sigs$plus_minus == "-1",]))

bp5  <- ggplot(sigs, aes(x=factor(1),y=plus_minus, fill=plus_minus)) +
        geom_bar(width = 1, stat = "identity", show.legend = TRUE) +
        scale_fill_manual(values = c("lightskyblue","indianred")) + labs(NULL) +
        coord_polar("y", start=0) + my_theme
bp5

### Characterized sigs bar chart
sigs_pval_lfc <- sigs %>% filter(cat == "characterized") %>% filter(abs(log2FoldChange)>1.5) %>% 
  arrange(desc(log2FoldChange))
sigs_pval_lfc$gene <- factor(sigs_pval_lfc$gene, ordered = TRUE) 

# Keep the X number of genes (up/down)
num_keep = 10
if(nrow(sigs_pval_lfc)>20){num_keep <- 10} else {num_keep <- nrow(sigs_pval_lfc)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(sigs_pval_lfc)-num_keep+1), nrow(sigs_pval_lfc)))

bp6 <- ggplot(sigs_pval_lfc[rows_keep,], aes(x=as.numeric(row.names(sigs_pval_lfc[rows_keep,])), 
                                             y=log2FoldChange, fill=plus_minus)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             scale_fill_manual(values=c("salmon1","royalblue1")) +
             scale_y_continuous(breaks = seq(-4, 4, by = 2), limits = c(-5.5,5.5)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp6

bp7 <- ggplot(sigs_characterized, aes(x=as.numeric(row.names(sigs_characterized)), 
                                             y=log2FoldChange, fill=feature)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             # scale_fill_manual(values=c("salmon1","royalblue1")) +
             # scale_y_continuous(breaks = seq(-4, 8, by = 2), limits = c(-5.5,9)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp7

# volcano plot pval/lfc
p9 <- ggplot(sigs, aes(x = log2FoldChange, y = -log10(pvalue), color = plus_minus)) +
             geom_point(show.legend = FALSE) +
             scale_color_manual(values = c("salmon1","royalblue1")) +
             scale_x_continuous(breaks = seq(-4, 4, by = 2), limits = c(-5.5,5.5)) + 
             scale_y_continuous(breaks = seq(4, 20, by = 4), limits = c(2.5,20)) + my_theme + 
             theme(axis.title.x=element_blank(),axis.title.y=element_blank())

# Output tables of DEGs
sigs$analysis <- "diploid_heat"
sigs_pval_lfc$analysis <- "diploid_heat"
write.table(sigs, file = "DESEQ2_output/diploid_heat/diploid_heat-sig-genes-apeglm-pval-only.csv", row.names = F)
write.table(sigs_pval_lfc, file = "DESEQ2_output/diploid_heat/diploid_heat-sig-genes-apeglm-pval-lfc.csv", row.names = F)

# Save plots
ggsave("DESEQ2_output/diploid_heat/diploid_heat_DEG_uncharacterized_pval_apeglm.png",
       plot   = bp4, bg = "transparent",
       dpi    = 2500,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_heat/diploid_heat-DEG-apeglm-updown-pie.png",
       plot   = bp5,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_heat/diploid_heat-DEG-apeglm-updown-bar.png",
       plot   = bp6,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 5,
       units  = "in")

ggsave("DESEQ2_output/diploid_heat/diploid_heat-DEG-apeglm.png",
       plot   = bp7,
       dpi    = 600,
       device = "png",
       width  = 5,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_heat/diploid_heat-volcano_pval_genes_apeglm.png",
       plot   = p9,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

beep(2)
```

# DIPLOID only - effect of desiccation
```{r DIPLOID:DESICCATION, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "A" | group == "E")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ condition)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across 1/3 of the samples - https://support.bioconductor.org/p/110307/
keep <- rowSums(DESeq2::counts(dds) >= 10) >= ncol(cts_trt)/3
dds <- dds[keep,]

# Generate Contrasts
contrast_list    <- c("condition", "desiccation", "control") # factor, treatment group, control
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "DESEQ2_output/diploid_desiccation/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "diploid_desiccation"
# all_genes_pvalue    <- cbind(gene=all_genes$gene,padj=all_genes$padj)

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "diploid_desiccation"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
# all_genes_apeglm           <- left_join(all_genes_apeglm, as.data.frame(all_genes_pvalue), by = "gene")
all_genes_apeglm$treatment <- "diploid_desiccation"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "diploid_desiccation"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "diploid_desiccation" 

sig_genes_normal <- res_table_normal %>%
  filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "diploid_desiccation" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "diploid_desiccation" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "diploid_desiccation" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05","DEG_normal-p0.05",
                            "DEG_apeglm-p0.05","DEG_ashr-p0.05")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-gene-counts.csv",        row.names = TRUE)

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)

bp0 <- biplot(p, showLoadings = FALSE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

bp0

bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'group',
              colkey = c('A' = 'royalblue1' ,
                         'B' = 'green1'     ,
                         'C' = 'yellow2' ,
                         'D' = 'orange' ,
                         'E' = 'orangered1',
                         'F' = 'red3'),
              shape = 'condition',
              shapekey = c('control' = 19, 'desiccation' = 19),
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              ellipseFillKey = c('control' = 'royalblue1', 'desiccation' = 'red1'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-150,180), ylim = c(-50, 70),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) +
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1
# 
bp2 <- pairsplot(p,
                components = getComponents(p, c(1:5)),
                triangle = TRUE, trianglelabSize = 12,
                hline = 0, vline = 0,
                pointSize = 1,
                gridlines.major = FALSE, gridlines.minor = FALSE,
                colby = 'condition',
                title = 'Pairs plot', plotaxes = FALSE,
                margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2

# Plot pheatmap
rld <- rlog(dds, blind=T)
rld_mat <- assay(rld)
pca <- prcomp(t(rld_mat))
rld_cor <- cor(rld_mat)
bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])

bp3
 
 
setwd("DESEQ2_output/diploid_desiccation")

ggexport(filename = "diploid_desiccation-BIPLOT.png",
         plot   = bp0,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "diploid_desiccation-PCA.png",
         plot   = bp1,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "diploid_desiccation-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

ggexport(filename = "diploid_desiccation-pheatmap.png",
         plot   = bp3,
         res    = 600,
         width  = 6000,
         height = 6000)

setwd("..")
setwd("..")
 
# Gene expression heatmap
# Useful: https://www.datanovia.com/en/lessons/heatmap-in-r-static-and-interactive-visualization/#:~:text=%2C%20Colv%20%3D%20Colv)-,Complex%20heatmap,different%20data%20from%20different%20sources.
# From: https://github.com/mousepixels/sanbomics/blob/main/tutorial_complex_Heatmap.Rmd

# Grab all results from DESeq2 results table
all <- na.omit(res_table_apeglm)
sigs <- all[all$padj < 0.05,]
# sigs <- all[all$baseMean > 50,]
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')
# filter by uncharacterized (remove for top-20 graph)
grx <- glob2rx("unchar*") 
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(grx, description)))
sigs_characterized   <- anti_join(sigs, sigs_uncharacterized, by = 'gene')

# Get list of genes and associated metadata
df <- as.data.frame(sigs_characterized)
row.names(df) <- df$gene
df_sigs <- as.data.frame(sigs)
row.names(df_sigs) <- df_sigs$gene
df_all <- as.data.frame(all)
row.names(df_all) <- df_all$gene
df.top <- df[ (df$baseMean > 10) & (abs(df$log2FoldChange) > 1.5),]
df.top <- df.top[order(df.top$log2FoldChange, decreasing = TRUE),]
# head(df.top)

# Get normalized count data from dds object
rlog_out <- rlog(dds, blind=FALSE) 
mat     <- assay(rlog_out)[rownames(df.top), rownames(coldata_trt)] #sig genes x samples
colnames(mat) <- rownames(coldata_trt)
base_mean <- rowMeans(mat)
mat_all <- assay(rlog_out)[rownames(df_all), rownames(coldata_trt)]
colnames(mat_all) <- rownames(coldata_trt)
mat_sigs <- assay(rlog_out)[rownames(df_sigs), rownames(coldata_trt)]
colnames(mat_sigs) <- rownames(coldata_trt)

# Generate Z-score
mat.scaled <- t(apply(mat, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled)<-colnames(mat)
mat.scaled_all <- t(apply(mat_all, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_all)<-colnames(mat_all)
mat.scaled_sigs <- t(apply(mat_sigs, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_sigs)<-colnames(mat_sigs)

# Keep the X number of  
if(nrow(df.top)>20){num_keep <- 10} else {num_keep <- nrow(df.top)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(mat.scaled)-num_keep), nrow(mat.scaled)) )
l2_val <- as.matrix(df.top[rows_keep,]$log2FoldChange) #getting log2 value for each gene we are keeping
colnames(l2_val)<-"logFC"
mean <- as.matrix(df.top[rows_keep,]$baseMean) #getting mean value for each gene we are keeping
colnames(mean)<-"AveExpr"

#maps values between b/w/r for min and max l2 values
col_logFC <- colorRamp2(c(min(l2_val),0, max(l2_val)), c("blue", "white", "red"))

#maps between 0% quantile, and 75% quantile of mean values --- 0, 25, 50, 75, 100
col_AveExpr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white", "red"))

LOCID <- as.data.frame(rownames(mat.scaled[rows_keep,]))
colnames(LOCID) <- 'gene'
LOCID <- left_join(LOCID, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_complete <- as.data.frame(rownames(mat.scaled))
colnames(LOCID_complete) <- 'gene'
LOCID_complete <- left_join(LOCID_complete, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_sigs <- as.data.frame(rownames(mat.scaled_sigs))
colnames(LOCID_sigs) <- 'gene'
LOCID_sigs <- left_join(LOCID_sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

# Plot heatmap + annotations of top X number of DEGs (w/ lfc and pvalue cutoff)
png(filename = "DESEQ2_output/diploid_desiccation/diploid_desiccation_gene_heatmap.png", 
    res = 600, width = 6000, height = 4000, units = "px", pointsize = 6)
# Z-score matrix
h1 <- Heatmap(mat.scaled[rows_keep,], 
            column_labels       = colnames(mat.scaled),
            row_labels          = LOCID$description,
            row_names_side      = "left",
            row_names_gp        = gpar(fontsize = 8),
            row_names_centered  = F,
            cluster_columns     = F,
            cluster_rows        = F,
            row_names_max_width =  unit(10, "cm"))

# log2FoldChange matrix
h2 <- Heatmap(l2_val, row_labels = df.top$gene[rows_keep], 
            cluster_rows = F, 
            name="logFC", 
            col = col_logFC,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(l2_val[i, j],1), x, y)
            })

# Average expression
h3 <- Heatmap(mean, row_labels = df.top$symbol[rows_keep], 
            cluster_rows = F, 
            name = "AveExpr", 
            col=col_AveExpr,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(mean[i, j],0), x, y)
            })
h<-h1+h2+h3
h
# par(mar=c(5,8,4,1)+.1)
print(h)
dev.off()

# Plot heatmap
png(filename = "DESEQ2_output/diploid_desiccation/diploid_desiccation_heatmap_pval_genes_apeglm.png", 
    res = 1200, width = 6000, height = 6000, units = "px", pointsize = 6)
# par(mar=c(5,8,4,1)+.1)
col <- colorRampPalette(brewer.pal(11, "RdYlBu"))(256)
heatmap(mat.scaled_sigs, scale = "none", col =  col, 
        ColSideColors = c(rep("royalblue1", 11), rep("orangered1", 10)))
dev.off()

# Volcano plots
# All genes
p1 <- ggplot(all_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p2 <- ggplot(all_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p3 <- ggplot(all_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p4 <- ggplot(all_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

# Significant Genes
p5 <- ggplot(sig_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p6 <- ggplot(sig_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p7 <- ggplot(sig_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p8 <- ggplot(sig_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

# Save plots
ggsave("DESEQ2_output/diploid_desiccation/Volcano_all_genes.png",
       plot   = p1,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_desiccation/Volcano_all_genes_normal.png",
       plot   = p2,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_desiccation/Volcano_all_genes_ashr.png",
       plot   = p3,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_desiccation/Volcano_all_genes_apeglm.png",
       plot   = p4,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_desiccation/Volcano_pval_lfc_genes.png",
       plot   = p5,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_desiccation/Volcano_pval_lfc_genes_normal.png",
       plot   = p6,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_desiccation/Volcano_pval_lfc_genes_ashr.png",
       plot   = p7,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_desiccation/Volcano_pval_lfc_genes_apeglm.png",
       plot   = p8,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

### Pie chart of unknown vs known DEG
# Prepare dataframe for plot
all <- na.omit(res_table_apeglm) # Grab all DEGs, using apeglm shrinkage estimator
sigs <- all[all$padj < 0.05,] # filter for significant DEGs
# sigs <- all[all$baseMean > 50,]
# sigs <- sigs[abs(sigs$log2FoldChange)>1.5,]
sigs$cat <- "characterized"
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene') # add information from gene master table
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(glob2rx("*unchar*"), description))) # find uncharacterized DEGs
if(nrow(sigs_uncharacterized) > 0){sigs_uncharacterized$cat <- "uncharacterized"} # add category label
sigs <- anti_join(sigs, sigs_uncharacterized,  by = "gene") %>% bind_rows(sigs_uncharacterized) # merge unchar and characterized

sigs$count <- 1
bp4  <- ggplot(sigs, aes(x=factor(1),y=count, fill=cat)) +
        geom_bar(width = 1, stat = "identity", show.legend = FALSE) +
        scale_fill_manual(values = c("gray12",'azure4')) + labs(NULL) +
        coord_polar("y", start=0) + my_theme + theme(axis.title.x=element_blank(),
                                                     axis.title.y=element_blank(),
                                                     axis.text.x=element_blank(),
                                                     axis.text.y=element_blank(),
                                                     axis.ticks.y=element_blank())
bp4

### Pie chart of unknown of up or down regulated
sigs$plus_minus <- as.character(sign(sigs$log2FoldChange))

bp5  <- ggplot(sigs, aes(x=factor(1),y=plus_minus, fill=plus_minus)) +
        geom_bar(width = 1, stat = "identity", show.legend = FALSE) +
        scale_fill_manual(values = c("lightskyblue","indianred")) + labs(NULL) +
        coord_polar("y", start=0) + my_theme
bp5

### Characterized sigs bar chart
sigs_pval_lfc <- sigs %>% filter(cat == "characterized") %>% filter(abs(log2FoldChange)>1.5) %>% 
  arrange(desc(log2FoldChange))
sigs_pval_lfc$gene <- factor(sigs_pval_lfc$gene, ordered = TRUE) 

# Keep the X number of genes (up/down)
num_keep = 10
if(nrow(sigs_pval_lfc)>20){num_keep <- 10} else {num_keep <- nrow(sigs_pval_lfc)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(sigs_pval_lfc)-num_keep+1), nrow(sigs_pval_lfc)))

bp6 <- ggplot(sigs_pval_lfc[rows_keep,], aes(x=as.numeric(row.names(sigs_pval_lfc[rows_keep,])), 
                                             y=log2FoldChange, fill=plus_minus)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             scale_fill_manual(values=c("salmon1","royalblue1")) +
             scale_y_continuous(breaks = seq(-4, 8, by = 2), limits = c(-5.5,9)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp6


bp7 <- ggplot(sigs_characterized, aes(x=as.numeric(row.names(sigs_characterized)), 
                                             y=log2FoldChange, fill=feature)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             # scale_fill_manual(values=c("salmon1","royalblue1")) +
             # scale_y_continuous(breaks = seq(-4, 8, by = 2), limits = c(-5.5,9)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp7

# volcano plot pval/lfc
p9 <- ggplot(sigs, aes(x = log2FoldChange, y = -log10(pvalue), color = plus_minus)) +
             geom_point(show.legend = FALSE) +
             scale_color_manual(values = c("salmon1","royalblue1")) +
             scale_x_continuous(breaks = seq(-4, 4, by = 2), limits = c(-5.5,5.5)) + 
             scale_y_continuous(breaks = seq(4, 20, by = 4), limits = c(2.5,20)) + my_theme + 
             theme(axis.title.x=element_blank(),axis.title.y=element_blank())

# Output tables of DEGs
sigs$analysis <- "diploid_desiccation"
sigs_pval_lfc$analysis <- "diploid_desiccation"
write.table(sigs, file = "DESEQ2_output/diploid_desiccation/diploid_desiccation-sig-genes-apeglm-pval-only.csv", row.names = F)
write.table(sigs_pval_lfc, file = "DESEQ2_output/diploid_desiccation/diploid_desiccation-sig-genes-apeglm-pval-lfc.csv", row.names = F)

# Save plots
ggsave("DESEQ2_output/diploid_desiccation/diploid_desiccation-DEG-uncharacterized-pval-apeglm.png",
       plot   = bp4, bg = "transparent",
       dpi    = 1200,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_desiccation/diploid_desiccation-DEG-apeglm-updown-pie.png",
       plot   = bp5,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_desiccation/diploid_desiccation-DEG-apeglm-updown-bar.png",
       plot   = bp6,
       dpi    = 600,
       device = "png",
       width  = 5,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_desiccation/diploid_desiccation-DEG-apeglm.png",
       plot   = bp7,
       dpi    = 600,
       device = "png",
       width  = 5,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/diploid_desiccation/diploid_desiccation-volcano_pval_genes_apeglm.png",
       plot   = p9,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")


beep(2)
```

# TRIPLOID only - effect of temperature
```{r TRIPLOID:HEAT, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "B" | group == "D")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ condition)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across 1/3 of the samples - https://support.bioconductor.org/p/110307/
keep <- rowSums(DESeq2::counts(dds) >= 10) >= ncol(cts_trt)/3
dds <- dds[keep,]

# Generate Contrasts
contrast_list    <- c("condition", "heat", "control") # factor, treatment group, control
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "DESEQ2_output/triploid_heat/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "triploid_heat"

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "triploid_heat"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
all_genes_apeglm$treatment <- "triploid_heat"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "triploid_heat"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "triploid_heat" 

sig_genes_normal <- res_table_normal %>%
  filter(padj < padj.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "triploid_heat" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "triploid_heat" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "triploid_heat" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05","DEG_normal-p0.05",
                            "DEG_apeglm-p0.05","DEG_ashr-p0.05")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "DESEQ2_output/triploid_heat/TRIPLOID_HEAT-gene-counts.csv",        row.names = TRUE)

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)

bp0 <- biplot(p, showLoadings = FALSE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

bp0
 
bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'group',
              colkey = c('A' = 'royalblue1' ,
                         'B' = 'green1'     ,
                         'C' = 'yellow2' ,
                         'D' = 'orange' ,
                         'E' = 'orangered1',
                         'F' = 'red3'),
              shape = 'condition',
              shapekey = c('control' = 17, 'heat' = 17),
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              ellipseFillKey = c('control' = 'green1', 'heat' = 'orangered3'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-160,140), ylim = c(-100, 80),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) +
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1
# 
bp2 <- pairsplot(p,
                components = getComponents(p, c(1:5)),
                triangle = TRUE, trianglelabSize = 12,
                hline = 0, vline = 0,
                pointSize = 1,
                gridlines.major = FALSE, gridlines.minor = FALSE,
                colby = 'condition',
                title = 'Pairs plot', plotaxes = FALSE,
                margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2
 
# Plot pheatmap
rld <- rlog(dds, blind=T)
rld_mat <- assay(rld)
pca <- prcomp(t(rld_mat))
rld_cor <- cor(rld_mat)
bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])

bp3
 
 
setwd("DESEQ2_output/triploid_heat")

ggexport(filename = "triploid_heat-BIPLOT.png",
         plot   = bp0,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "triploid_heat-PCA.png",
         plot   = bp1,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "triploid_heat-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

ggexport(filename = "triploid_heat-pheatmap.png",
         plot   = bp3,
         res    = 600,
         width  = 6000,
         height = 6000)

setwd("..")
setwd("..")
 
# Gene expression heatmap
# Useful: https://www.datanovia.com/en/lessons/heatmap-in-r-static-and-interactive-visualization/#:~:text=%2C%20Colv%20%3D%20Colv)-,Complex%20heatmap,different%20data%20from%20different%20sources.
# From: https://github.com/mousepixels/sanbomics/blob/main/tutorial_complex_Heatmap.Rmd

# Grab all results from DESeq2 results table
all <- na.omit(res_table_apeglm)
sigs <- all[all$padj < 0.05,]
# sigs <- all[all$baseMean > 50,]
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')
# filter by uncharacterized (remove for top-20 graph)
grx <- glob2rx("unchar*") 
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(grx, description)))
sigs_characterized   <- anti_join(sigs, sigs_uncharacterized, by = 'gene')

# Get list of genes and associated metadata
df <- as.data.frame(sigs_characterized)
row.names(df) <- df$gene
df_sigs <- as.data.frame(sigs)
row.names(df_sigs) <- df_sigs$gene
df_all <- as.data.frame(all)
row.names(df_all) <- df_all$gene
df.top <- df[ (df$baseMean > 10) & (abs(df$log2FoldChange) > 1.5),]
df.top <- df.top[order(df.top$log2FoldChange, decreasing = TRUE),]
# head(df.top)

# Get normalized count data from dds object
rlog_out <- rlog(dds, blind=FALSE) 
mat     <- assay(rlog_out)[rownames(df.top), rownames(coldata_trt)] #sig genes x samples
colnames(mat) <- rownames(coldata_trt)
base_mean <- rowMeans(mat)
mat_all <- assay(rlog_out)[rownames(df_all), rownames(coldata_trt)]
colnames(mat_all) <- rownames(coldata_trt)
mat_sigs <- assay(rlog_out)[rownames(df_sigs), rownames(coldata_trt)]
colnames(mat_sigs) <- rownames(coldata_trt)

# Generate Z-score
mat.scaled <- t(apply(mat, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled)<-colnames(mat)
mat.scaled_all <- t(apply(mat_all, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_all)<-colnames(mat_all)
mat.scaled_sigs <- t(apply(mat_sigs, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_sigs)<-colnames(mat_sigs)

# Keep the X number of  
if(nrow(df.top)>20){num_keep <- 10} else {num_keep <- nrow(df.top)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(mat.scaled)-num_keep), nrow(mat.scaled)) )
l2_val <- as.matrix(df.top[rows_keep,]$log2FoldChange) #getting log2 value for each gene we are keeping
colnames(l2_val)<-"logFC"
mean <- as.matrix(df.top[rows_keep,]$baseMean) #getting mean value for each gene we are keeping
colnames(mean)<-"AveExpr"

#maps values between b/w/r for min and max l2 values
col_logFC <- colorRamp2(c(min(l2_val),0, max(l2_val)), c("blue", "white", "red"))

#maps between 0% quantile, and 75% quantile of mean values --- 0, 25, 50, 75, 100
col_AveExpr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white", "red"))

LOCID <- as.data.frame(rownames(mat.scaled[rows_keep,]))
colnames(LOCID) <- 'gene'
LOCID <- left_join(LOCID, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_complete <- as.data.frame(rownames(mat.scaled))
colnames(LOCID_complete) <- 'gene'
LOCID_complete <- left_join(LOCID_complete, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_sigs <- as.data.frame(rownames(mat.scaled_sigs))
colnames(LOCID_sigs) <- 'gene'
LOCID_sigs <- left_join(LOCID_sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

# Plot heatmap + annotations of top X number of DEGs (w/ lfc and pvalue cutoff)
png(filename = "DESEQ2_output/triploid_heat/triploid_heat_gene_heatmap.png", 
    res = 600, width = 6000, height = 4000, units = "px", pointsize = 6)
# Z-score matrix
h1 <- Heatmap(mat.scaled[rows_keep,], 
            column_labels       = colnames(mat.scaled),
            row_labels          = LOCID$description,
            row_names_side      = "left",
            row_names_gp        = gpar(fontsize = 8),
            row_names_centered  = F,
            cluster_columns     = F,
            cluster_rows        = F,
            row_names_max_width =  unit(10, "cm"))

# log2FoldChange matrix
h2 <- Heatmap(l2_val, row_labels = df.top$gene[rows_keep], 
            cluster_rows = F, 
            name="logFC", 
            col = col_logFC,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(l2_val[i, j],1), x, y)
            })

# Average expression
h3 <- Heatmap(mean, row_labels = df.top$symbol[rows_keep], 
            cluster_rows = F, 
            name = "AveExpr", 
            col=col_AveExpr,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(mean[i, j],0), x, y)
            })
h<-h1+h2+h3
h
# par(mar=c(5,8,4,1)+.1)
print(h)
dev.off()

# Plot heatmap
png(filename = "DESEQ2_output/triploid_heat/triploid_heat_heatmap_pval_genes_apeglm.png", 
    res = 1200, width = 6000, height = 6000, units = "px", pointsize = 6)
# par(mar=c(5,8,4,1)+.1)
col <- colorRampPalette(brewer.pal(11, "RdYlBu"))(256)
heatmap(mat.scaled_sigs, scale = "none", col =  col, 
        ColSideColors = c(rep("royalblue1", 11), rep("orangered1", 10)))
dev.off()

# Volcano plots
# All genes
p1 <- ggplot(all_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p2 <- ggplot(all_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p3 <- ggplot(all_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p4 <- ggplot(all_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

# Significant Genes
p5 <- ggplot(sig_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p6 <- ggplot(sig_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p7 <- ggplot(sig_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p8 <- ggplot(sig_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

# Save plots
ggsave("DESEQ2_output/triploid_heat/Volcano_all_genes.png",
       plot   = p1,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_heat/Volcano_all_genes_normal.png",
       plot   = p2,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_heat/Volcano_all_genes_ashr.png",
       plot   = p3,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_heat/Volcano_all_genes_apeglm.png",
       plot   = p4,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_heat/Volcano_pval_lfc_genes.png",
       plot   = p5,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_heat/Volcano_pval_lfc_genes_normal.png",
       plot   = p6,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_heat/Volcano_pval_lfc_genes_ashr.png",
       plot   = p7,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_heat/Volcano_pval_lfc_genes_apeglm.png",
       plot   = p8,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

### Pie chart of unknown vs known DEG
# Prepare dataframe for plot
all <- na.omit(res_table_apeglm) # Grab all DEGs, using apeglm shrinkage estimator
sigs <- all[all$padj < 0.05,] # filter for significant DEGs
# sigs <- all[all$baseMean > 50,]
# sigs <- sigs[abs(sigs$log2FoldChange)>1.5,]
sigs$cat <- "characterized"
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene') # add information from gene master table
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(glob2rx("*unchar*"), description))) # find uncharacterized DEGs
if(nrow(sigs_uncharacterized) > 0){sigs_uncharacterized$cat <- "uncharacterized"} # add category label
sigs <- anti_join(sigs, sigs_uncharacterized,  by = "gene") %>% bind_rows(sigs_uncharacterized) # merge unchar and characterized

sigs$count <- 1
bp4  <- ggplot(sigs, aes(x=factor(1),y=count, fill=cat)) +
        geom_bar(width = 1, stat = "identity", show.legend = FALSE) +
        scale_fill_manual(values = c("gray12",'azure4')) + labs(NULL) +
        coord_polar("y", start=0) + my_theme + theme(axis.title.x=element_blank(),
                                                     axis.title.y=element_blank(),
                                                     axis.text.x=element_blank(),
                                                     axis.text.y=element_blank(),
                                                     axis.ticks.y=element_blank())
bp4

### Pie chart of unknown of up or down regulated
sigs$plus_minus <- as.character(sign(sigs$log2FoldChange))

bp5  <- ggplot(sigs, aes(x=factor(1),y=plus_minus, fill=plus_minus)) +
        geom_bar(width = 1, stat = "identity", show.legend = FALSE) +
        scale_fill_manual(values = c("lightskyblue","indianred")) + labs(NULL) +
        coord_polar("y", start=0) + my_theme
bp5

### Characterized sigs bar chart
sigs_pval_lfc <- sigs %>% filter(cat == "characterized") %>% filter(abs(log2FoldChange)>1.5) %>% 
  arrange(desc(log2FoldChange))
sigs_pval_lfc$gene <- factor(sigs_pval_lfc$gene, ordered = TRUE) 

# Keep the X number of genes (up/down)
num_keep = 10
if(nrow(sigs_pval_lfc)>20){num_keep <- 10} else {num_keep <- nrow(sigs_pval_lfc)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(sigs_pval_lfc)-num_keep+1), nrow(sigs_pval_lfc)))

bp6 <- ggplot(sigs_pval_lfc[rows_keep,], aes(x=as.numeric(row.names(sigs_pval_lfc[rows_keep,])), 
                                             y=log2FoldChange, fill=plus_minus)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             scale_fill_manual(values=c("salmon1","royalblue1")) +
             scale_y_continuous(breaks = seq(-5, 5, by = 2), limits = c(-5,5)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp6


bp7 <- ggplot(sigs_characterized, aes(x=as.numeric(row.names(sigs_characterized)), 
                                             y=log2FoldChange, fill=feature)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             # scale_fill_manual(values=c("salmon1","royalblue1")) +
             # scale_y_continuous(breaks = seq(-4, 8, by = 2), limits = c(-5.5,9)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp7

# volcano plot pval/lfc
p9 <- ggplot(sigs, aes(x = log2FoldChange, y = -log10(pvalue), color = plus_minus)) +
             geom_point(show.legend = FALSE) +
             scale_color_manual(values = c("salmon1","royalblue1")) +
             scale_x_continuous(breaks = seq(-4, 4, by = 2), limits = c(-5.5,5.5)) + 
             scale_y_continuous(breaks = seq(4, 20, by = 4), limits = c(2.5,20)) + my_theme + 
             theme(axis.title.x=element_blank(),axis.title.y=element_blank())

# Output tables of DEGs
sigs$analysis <- "triploid_heat"
sigs_pval_lfc$analysis <- "triploid_heat"
write.table(sigs, file = "DESEQ2_output/triploid_heat/triploid_heat-sig-genes-apeglm-pval-only.csv", row.names = F)
write.table(sigs_pval_lfc, file = "DESEQ2_output/triploid_heat/triploid_heat-sig-genes-apeglm-pval-lfc.csv", row.names = F)

# Save plots
ggsave("DESEQ2_output/triploid_heat/triploid_heat-DEG-uncharacterized-pval-apeglm.png",
       plot   = bp4, bg = "transparent",
       dpi    = 1200,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_heat/triploid_heat-DEG-apeglm-updown-pie.png",
       plot   = bp5,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_heat/triploid_heat-DEG-apeglm-updown-bar.png",
       plot   = bp6,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_heat/triploid_heat-DEG-apeglm.png",
       plot   = bp7,
       dpi    = 600,
       device = "png",
       width  = 5,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_heat/triploid_heat-volcano_pval_genes_apeglm.png",
       plot   = p9,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")


beep(2)
```

# TRIPLOID only - effect of desiccation
```{r TRIPLOID:DESICCATION, warning=FALSE, include=TRUE}

# A - diploid; control
# B - triploid; control
# C - diploid; heat
# D - triploid; heat
# E - diploid; desiccation
# F - triploid; desiccation

# Filter data
coldata_trt <- coldata %>% filter(group == "B" | group == "F")
cts_trt     <- subset(cts, select=row.names(coldata_trt))

# Calculate DESeq object
dds <- DESeqDataSetFromMatrix(countData = cts_trt,
                              colData = coldata_trt,
                              design = ~ condition)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients

# Filtering: keep genes that have at least 10 counts across 1/3 of the samples - https://support.bioconductor.org/p/110307/
keep <- rowSums(DESeq2::counts(dds) >= 10) >= ncol(cts_trt)/3
dds <- dds[keep,]

# Generate Contrasts
contrast_list    <- c("condition", "desiccation", "control") # factor, treatment group, control
res_table        <- results(dds, contrast=contrast_list, alpha = 0.05)
res_table_normal <- lfcShrink(dds,
                              coef=2, 
                              type="normal") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_apeglm <- lfcShrink(dds,
                              coef=2, 
                              type="apeglm") # lfcThreshold = 0.585)  # a lfc threshold of 1 = 2-fold change, 0.585 = 1.5-fold change
res_table_ashr   <- lfcShrink(dds,
                              coef=2, 
                              type="ashr")

# Plot MA
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)

pdf(file= "DESEQ2_output/triploid_desiccation/MA_plots.pdf" )
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
DESeq2::plotMA(res_table_normal, xlim=xlim, ylim=ylim, main="normal", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_apeglm, xlim=xlim, ylim=ylim, main="apeglm", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
DESeq2::plotMA(res_table_ashr, xlim=xlim, ylim=ylim, main="ashr", cex=.8)
abline(h=c(-1.5,1.5), col="dodgerblue", lwd=2)
dev.off()

# Set cut.offs
padj.cutoff <- 0.05 # pvalues not produced with apeglm shrinkage estimator
svalue.cutoff <- 0.005 # 0.005 corresponds to pvalue of 0.05
lfc.cutoff <- 1.5 # fold-change cutoff, 1.5 or 2-fold is recommended

# Rearrange tables by gene
res_table <- res_table %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_normal <- res_table_normal %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_apeglm <- res_table_apeglm %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

res_table_ashr <- res_table_ashr %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>%
  as_tibble()

# How many genes are differentially expressed compared to Control?
colnames(LOCID_Roslin_gigas_gene_mito_table) <- c("gene","description","feature")
all_genes           <- left_join(res_table, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes           <- all_genes %>% arrange(-log2FoldChange)
all_genes$treatment <- "triploid_desiccation"
# all_genes_pvalue    <- cbind(gene=all_genes$gene,padj=all_genes$padj)

all_genes_normal           <- left_join(res_table_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_normal           <- all_genes_normal %>% arrange(-log2FoldChange)
all_genes_normal$treatment <- "triploid_desiccation"

all_genes_apeglm           <- left_join(res_table_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_apeglm           <- all_genes_apeglm %>% arrange(-log2FoldChange)
# all_genes_apeglm           <- left_join(all_genes_apeglm, as.data.frame(all_genes_pvalue), by = "gene")
all_genes_apeglm$treatment <- "triploid_desiccation"

all_genes_ashr           <- left_join(res_table_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
all_genes_ashr           <- all_genes_ashr %>% arrange(-log2FoldChange)
all_genes_ashr$treatment <- "triploid_desiccation"

# Filter results by
sig_genes <- res_table %>%
  filter(padj < padj.cutoff)
sig_genes <- left_join(sig_genes, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes <- sig_genes %>% arrange(-log2FoldChange)
sig_genes$treatment <- "triploid_desiccation" 

sig_genes_normal <- res_table_normal %>%
  filter(abs(log2FoldChange) > lfc.cutoff)
sig_genes_normal <- left_join(sig_genes_normal, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_normal <- sig_genes_normal %>% arrange(-log2FoldChange)
sig_genes_normal$treatment <- "triploid_desiccation" 

sig_genes_apeglm <- res_table_apeglm %>%
  filter(padj < padj.cutoff)
sig_genes_apeglm <- left_join(sig_genes_apeglm, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_apeglm <- sig_genes_apeglm %>% arrange(-log2FoldChange)
sig_genes_apeglm$treatment <- "triploid_desiccation" 

sig_genes_ashr <- res_table_ashr %>%
  filter(padj < padj.cutoff)
sig_genes_ashr <- left_join(sig_genes_ashr, LOCID_Roslin_gigas_gene_mito_table, by = "gene")
sig_genes_ashr <- sig_genes_ashr %>% arrange(-log2FoldChange)
sig_genes_ashr$treatment <- "triploid_desiccation" 

# Save single gene expression counts to a data frame object, for each shrinkage estimator
single_gene_expression_unshrunk <- data.frame()
for (i in 1:nrow(sig_genes)){
  LOCID <- sig_genes$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_unshrunk <- rbind(single_gene_expression_unshrunk,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_normal <- data.frame()
for (i in 1:nrow(sig_genes_normal)){
  LOCID <- sig_genes_normal$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_normal <- rbind(single_gene_expression_normal,gene_counts_data)
  rm(gene_counts_data)
}


single_gene_expression_apeglm <- data.frame()
for (i in 1:nrow(sig_genes_apeglm)){
  LOCID <- sig_genes_apeglm$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_apeglm <- rbind(single_gene_expression_apeglm,gene_counts_data)
  rm(gene_counts_data)
}

single_gene_expression_ashr <- data.frame()
for (i in 1:nrow(sig_genes_ashr)){
  LOCID <- sig_genes_ashr$gene[i]
  gene_counts_data <- plotCounts(dds, gene=LOCID, intgroup="ploidy", returnData=TRUE)
  gene_counts_data$gene <- LOCID
  gene_counts_data$sample <- row.names(gene_counts_data)
  row.names(gene_counts_data) <- NULL
  single_gene_expression_ashr <- rbind(single_gene_expression_ashr,gene_counts_data)
  rm(gene_counts_data)
}

write.table(single_gene_expression_unshrunk, file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-unshrunk-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_normal, file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-normal-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_apeglm, file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-apeglm-single-gene-counts.csv", row.names = FALSE)
write.table(single_gene_expression_ashr, file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-ashr-single-gene-counts.csv", row.names = FALSE)

print(paste("# genes before filtering:", nrow(cts))) # count number of genes before filtering
print(paste("# genes after filtering:", length(dds))) # count number of genes after filtering
print(paste("# of genes dropped:", nrow(cts) - length(dds), sep=" ")) # count number of genes dropped
print(paste("# DEGs, all-genes:", nrow(all_genes)))
print(paste("# DEGs, unshrunken, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes)))
print(paste("# DEGs, normal shrinkage estimator, lfc = 1.5:", nrow(sig_genes_normal)))
print(paste("# DEGs, apeglm shrinkage estimator, svalue = 0.005, lfc = 1.5:", nrow(sig_genes_apeglm)))
print(paste("# DEGs, ashr shrinkage estimator, pvalue = 0.05, lfc = 1.5:", nrow(sig_genes_ashr)))
gene_counts <- as.data.frame(c(nrow(cts),length(dds),(nrow(cts)-length(dds)),nrow(all_genes),
                            nrow(all_genes_normal),nrow(all_genes_apeglm),nrow(all_genes_ashr),
                            nrow(sig_genes),nrow(sig_genes_normal),
                            nrow(sig_genes_apeglm),nrow(sig_genes_ashr)))
row.names(gene_counts) <- c("genes_before_filtering","genes_after_filtering","genes_dropped",
                            "DEGs_all-genes","DEGs_all-genes-normal",
                            "DEGs_all-genes-apeglm","DEGs_all-genes-ashr",
                            "DEG_unshrunken-p0.05","DEG_normal-p0.05",
                            "DEG_apeglm-p0.05","DEG_ashr-p0.05")
colnames(gene_counts) <- "count"

# Output tables of DEGs
write.table(all_genes,        file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-ALL-DEG.csv",            row.names = FALSE)
write.table(all_genes_normal, file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-ALL-DEG-normal.csv",     row.names = FALSE)
write.table(all_genes_apeglm, file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-ALL-DEG-apeglm.csv",     row.names = FALSE)
write.table(all_genes_ashr,   file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-ALL-DEG-ashr.csv",       row.names = FALSE)
write.table(sig_genes,        file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-SIG-DEG-unshrunken.csv", row.names = FALSE)
write.table(sig_genes_normal, file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-SIG-DEG-normal.csv",     row.names = FALSE)
write.table(sig_genes_apeglm, file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-SIG-DEG-apeglm.csv",     row.names = FALSE)
write.table(sig_genes_ashr,   file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-SIG-DEG-ashr.csv",       row.names = FALSE)
write.table(gene_counts,      file = "DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-gene-counts.csv",        row.names = TRUE)

# Plot PCA
p <- pca(assay(vst(dds)), metadata = coldata_trt)
screeplot(p, axisLabSize = 18, titleLabSize = 22)

bp0 <- biplot(p, showLoadings = FALSE, colby = 'group',
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

bp0

bp1 <- biplot(p, x = "PC1", y = "PC2",
              colby = 'group',
              colkey = c('A' = 'royalblue1' ,
                         'B' = 'green1'     ,
                         'C' = 'yellow2' ,
                         'D' = 'orange' ,
                         'E' = 'orangered1',
                         'F' = 'red3'),
              shape = 'condition',
              shapekey = c('control' = 17, 'desiccation' = 17),
              pointSize = 4,
              showLoadings = FALSE,
              lab = NULL,
              drawConnectors = FALSE,
              ellipse = TRUE,
              ellipseLevel = 0.95,
              ellipseFill = FALSE,
              ellipseFillKey = c('control' = 'green1', 'desiccation' = 'red3'),
              ellipseAlpha = 1/4,
              ellipseLineSize = 1,
              xlim = c(-150,130), ylim = c(-70, 60),
              gridlines.major = FALSE,
              gridlines.minor = FALSE,
              borderWidth = 1,
              legendPosition = 'top', legendLabSize = 16, legendIconSize = 6.0) +
              theme(axis.text.x       = element_text(size=16,color="black"),
                    axis.text.y       = element_text(size=16,color="black"),
                    axis.ticks.x      = element_line(color="black"),
                    axis.ticks.y      = element_line(color="black"))
bp1
# 
# 
bp2 <- pairsplot(p,
                components = getComponents(p, c(1:5)),
                triangle = TRUE, trianglelabSize = 12,
                hline = 0, vline = 0,
                pointSize = 1,
                gridlines.major = FALSE, gridlines.minor = FALSE,
                colby = 'condition',
                title = 'Pairs plot', plotaxes = FALSE,
                margingaps = unit(c(-0.01, -0.01, -0.01, -0.01), 'cm'))
bp2
# 
# Plot pheatmap
rld <- rlog(dds, blind=T)
rld_mat <- assay(rld)
pca <- prcomp(t(rld_mat))
rld_cor <- cor(rld_mat)
bp3 <- pheatmap(rld_cor, annotation = coldata_trt[,1:2])

bp3
 
 
setwd("DESEQ2_output/triploid_desiccation")

ggexport(filename = "triploid_desiccation-BIPLOT.png",
         plot   = bp0,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "triploid_desiccation-PCA.png",
         plot   = bp1,
         res    = 600,
         width  = 4000,
         height = 5000)

ggexport(filename = "triploid_desiccation-PAIRS.png",
         plot   = bp2,
         res    = 600,
         width  = 6000,
         height = 6000)

ggexport(filename = "triploid_desiccation-pheatmap.png",
         plot   = bp3,
         res    = 600,
         width  = 6000,
         height = 6000)

setwd("..")
setwd("..")
 
# Gene expression heatmap
# Useful: https://www.datanovia.com/en/lessons/heatmap-in-r-static-and-interactive-visualization/#:~:text=%2C%20Colv%20%3D%20Colv)-,Complex%20heatmap,different%20data%20from%20different%20sources.
# From: https://github.com/mousepixels/sanbomics/blob/main/tutorial_complex_Heatmap.Rmd

# Grab all results from DESeq2 results table
all <- na.omit(res_table_apeglm)
sigs <- all[all$padj < 0.05,]
# sigs <- all[all$baseMean > 50,]
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')
# filter by uncharacterized (remove for top-20 graph)
grx <- glob2rx("unchar*") 
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(grx, description)))
sigs_characterized   <- anti_join(sigs, sigs_uncharacterized, by = 'gene')

# Get list of genes and associated metadata
df <- as.data.frame(sigs_characterized)
row.names(df) <- df$gene
df_sigs <- as.data.frame(sigs)
row.names(df_sigs) <- df_sigs$gene
df_all <- as.data.frame(all)
row.names(df_all) <- df_all$gene
df.top <- df[ (df$baseMean > 10) & (abs(df$log2FoldChange) > 1.5),]
df.top <- df.top[order(df.top$log2FoldChange, decreasing = TRUE),]
# head(df.top)

# Get normalized count data from dds object
rlog_out <- rlog(dds, blind=FALSE) 
mat     <- assay(rlog_out)[rownames(df.top), rownames(coldata_trt)] #sig genes x samples
colnames(mat) <- rownames(coldata_trt)
base_mean <- rowMeans(mat)
mat_all <- assay(rlog_out)[rownames(df_all), rownames(coldata_trt)]
colnames(mat_all) <- rownames(coldata_trt)
mat_sigs <- assay(rlog_out)[rownames(df_sigs), rownames(coldata_trt)]
colnames(mat_sigs) <- rownames(coldata_trt)

# Generate Z-score
mat.scaled <- t(apply(mat, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled)<-colnames(mat)
mat.scaled_all <- t(apply(mat_all, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_all)<-colnames(mat_all)
mat.scaled_sigs <- t(apply(mat_sigs, 1, scale)) #center and scale each column (Z-score) then transpose
colnames(mat.scaled_sigs)<-colnames(mat_sigs)

# Keep the X number of  
if(nrow(df.top)>20){num_keep <- 10} else {num_keep <- nrow(df.top)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(mat.scaled)-num_keep), nrow(mat.scaled)) )
l2_val <- as.matrix(df.top[rows_keep,]$log2FoldChange) #getting log2 value for each gene we are keeping
colnames(l2_val)<-"logFC"
mean <- as.matrix(df.top[rows_keep,]$baseMean) #getting mean value for each gene we are keeping
colnames(mean)<-"AveExpr"

#maps values between b/w/r for min and max l2 values
col_logFC <- colorRamp2(c(min(l2_val),0, max(l2_val)), c("blue", "white", "red"))

#maps between 0% quantile, and 75% quantile of mean values --- 0, 25, 50, 75, 100
col_AveExpr <- colorRamp2(c(quantile(mean)[1], quantile(mean)[4]), c("white", "red"))

LOCID <- as.data.frame(rownames(mat.scaled[rows_keep,]))
colnames(LOCID) <- 'gene'
LOCID <- left_join(LOCID, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_complete <- as.data.frame(rownames(mat.scaled))
colnames(LOCID_complete) <- 'gene'
LOCID_complete <- left_join(LOCID_complete, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

LOCID_sigs <- as.data.frame(rownames(mat.scaled_sigs))
colnames(LOCID_sigs) <- 'gene'
LOCID_sigs <- left_join(LOCID_sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene')

# Plot heatmap + annotations of top X number of DEGs (w/ lfc and pvalue cutoff)
png(filename = "DESEQ2_output/triploid_desiccation/triploid_desiccation_gene_heatmap.png", 
    res = 600, width = 6000, height = 4000, units = "px", pointsize = 6)
# Z-score matrix
h1 <- Heatmap(mat.scaled[rows_keep,], 
            column_labels       = colnames(mat.scaled),
            row_labels          = LOCID$description,
            row_names_side      = "left",
            row_names_gp        = gpar(fontsize = 8),
            row_names_centered  = F,
            cluster_columns     = F,
            cluster_rows        = F,
            row_names_max_width =  unit(10, "cm"))

# log2FoldChange matrix
h2 <- Heatmap(l2_val, row_labels = df.top$gene[rows_keep], 
            cluster_rows = F, 
            name="logFC", 
            col = col_logFC,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(l2_val[i, j],1), x, y)
            })

# Average expression
h3 <- Heatmap(mean, row_labels = df.top$symbol[rows_keep], 
            cluster_rows = F, 
            name = "AveExpr", 
            col=col_AveExpr,
            cell_fun = function(j, i, x, y, w, h, col) { # add text to each grid
              grid.text(round(mean[i, j],0), x, y)
            })
h<-h1+h2+h3
h
# par(mar=c(5,8,4,1)+.1)
print(h)
dev.off()

# Plot heatmap
png(filename = "DESEQ2_output/triploid_desiccation/triploid_desiccation_heatmap_pval_genes_apeglm.png", 
    res = 1200, width = 6000, height = 6000, units = "px", pointsize = 6)
# par(mar=c(5,8,4,1)+.1)
col <- colorRampPalette(brewer.pal(11, "RdYlBu"))(256)
heatmap(mat.scaled_sigs, scale = "none", col =  col, 
        ColSideColors = c(rep("royalblue1", 11), rep("red3", 11)))
dev.off()

# Volcano plots
# All genes
p1 <- ggplot(all_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p2 <- ggplot(all_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p3 <- ggplot(all_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p4 <- ggplot(all_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

# Significant Genes
p5 <- ggplot(sig_genes, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p6 <- ggplot(sig_genes_normal, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p7 <- ggplot(sig_genes_ashr, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

p8 <- ggplot(sig_genes_apeglm, aes(x = log2FoldChange, y = -log10(pvalue))) +
             geom_point(aes(color=log2FoldChange>0)) + my_theme

# Save plots
ggsave("DESEQ2_output/triploid_desiccation/Volcano_all_genes.png",
       plot   = p1,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_desiccation/Volcano_all_genes_normal.png",
       plot   = p2,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_desiccation/Volcano_all_genes_ashr.png",
       plot   = p3,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_desiccation/Volcano_all_genes_apeglm.png",
       plot   = p4,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_desiccation/Volcano_pval_lfc_genes.png",
       plot   = p5,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_desiccation/Volcano_pval_lfc_genes_normal.png",
       plot   = p6,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_desiccation/Volcano_pval_lfc_genes_ashr.png",
       plot   = p7,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_desiccation/Volcano_pval_lfc_genes_apeglm.png",
       plot   = p8,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

### Pie chart of unknown vs known DEG
# Prepare dataframe for plot
all <- na.omit(res_table_apeglm) # Grab all DEGs, using apeglm shrinkage estimator
sigs <- all[all$padj < 0.05,] # filter for significant DEGs
# sigs <- all[all$baseMean > 50,]
# sigs <- sigs[abs(sigs$log2FoldChange)>1.5,]
sigs$cat <- "characterized"
sigs <- left_join(sigs, LOCID_Roslin_gigas_gene_mito_table, by = 'gene') # add information from gene master table
sigs_uncharacterized <- with(sigs, subset(sigs, subset = grepl(glob2rx("*unchar*"), description))) # find uncharacterized DEGs
if(nrow(sigs_uncharacterized) > 0){sigs_uncharacterized$cat <- "uncharacterized"} # add category label
sigs <- anti_join(sigs, sigs_uncharacterized,  by = "gene") %>% bind_rows(sigs_uncharacterized) # merge unchar and characterized

count_unchar <- as.numeric(nrow(sigs[sigs$cat == "uncharacterized",]))
count_char <- as.numeric(nrow(sigs[sigs$cat == "characterized",]))

sigs$count <- 1
bp4  <- ggplot(sigs, aes(x=factor(1),y=count, fill=cat)) +
        geom_bar(width = 1, stat = "identity", show.legend = FALSE) +
        scale_fill_manual(values = c("gray12",'azure4')) + labs(NULL) +
        coord_polar("y", start=0) + my_theme + theme(axis.title.x=element_blank(),
                                                     axis.title.y=element_blank(),
                                                     axis.text.x=element_blank(),
                                                     axis.text.y=element_blank(),
                                                     axis.ticks.y=element_blank())
bp4

### Pie chart of unknown of up or down regulated
sigs$plus_minus <- as.character(sign(sigs$log2FoldChange))
count_plus <- as.numeric(nrow(sigs[sigs$plus_minus == "1",]))
count_minus <- as.numeric(nrow(sigs[sigs$plus_minus == "-1",]))

bp5  <- ggplot(sigs, aes(x=factor(1),y=plus_minus, fill=plus_minus)) +
        geom_bar(width = 1, stat = "identity", show.legend = TRUE) +
        scale_fill_manual(values = c("lightskyblue","indianred")) + labs(NULL) +
        coord_polar("y", start=0) + my_theme
bp5

### Characterized sigs bar chart
sigs_pval_lfc <- sigs %>% filter(cat == "characterized") %>% filter(abs(log2FoldChange)>1.5) %>% 
  arrange(desc(log2FoldChange))
sigs_pval_lfc$gene <- factor(sigs_pval_lfc$gene, ordered = TRUE) 

# Keep the X number of genes (up/down)
num_keep = 10
if(nrow(sigs_pval_lfc)>20){num_keep <- 10} else {num_keep <- nrow(sigs_pval_lfc)/2}
rows_keep <- c(seq(1:num_keep), seq((nrow(sigs_pval_lfc)-num_keep+1), nrow(sigs_pval_lfc)))

bp6 <- ggplot(sigs_pval_lfc[rows_keep,], aes(x=as.numeric(row.names(sigs_pval_lfc[rows_keep,])), 
                                             y=log2FoldChange, fill=plus_minus)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             scale_fill_manual(values=c("salmon1","royalblue1")) +
             scale_y_continuous(breaks = seq(-4, 8, by = 2), limits = c(-5.5,9)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp6

bp7 <- ggplot(sigs_characterized, aes(x=as.numeric(row.names(sigs_characterized)), 
                                             y=log2FoldChange, fill=feature)) +
             geom_bar(width = 1, stat = "identity", position = "dodge", color="black", size = 0.3, show.legend = F) +
             # scale_fill_manual(values=c("salmon1","royalblue1")) +
             # scale_y_continuous(breaks = seq(-4, 8, by = 2), limits = c(-5.5,9)) +
             scale_x_continuous(breaks = NULL) + xlab("") +
             coord_flip() + my_theme
bp7

# volcano plot pval/lfc
p9 <- ggplot(sigs, aes(x = log2FoldChange, y = -log10(pvalue), color = plus_minus)) +
             geom_point(show.legend = FALSE) +
             scale_color_manual(values = c("salmon1","royalblue1")) +
             scale_x_continuous(breaks = seq(-4, 4, by = 2), limits = c(-5.5,5.5)) + 
             scale_y_continuous(breaks = seq(4, 20, by = 4), limits = c(2.5,20)) + my_theme + 
             theme(axis.title.x=element_blank(),axis.title.y=element_blank())

# Output tables of DEGs
sigs$analysis <- "triploid_desiccation"
sigs_pval_lfc$analysis <- "triploid_desiccation"
write.table(sigs, file = "DESEQ2_output/triploid_desiccation/triploid_desiccation-sig-genes-apeglm-pval-only.csv", row.names = F)
write.table(sigs_pval_lfc, file = "DESEQ2_output/triploid_desiccation/triploid_desiccation-sig-genes-apeglm-pval-lfc.csv", row.names = F)

# Save plots
ggsave("DESEQ2_output/triploid_desiccation/triploid_desiccation_DEG_uncharacterized_pval_apeglm.png",
       plot   = bp4, bg = "transparent",
       dpi    = 1500,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_desiccation/triploid_desiccation-DEG-apeglm-updown-pie.png",
       plot   = bp5,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_desiccation/triploid_desiccation-DEG-apeglm-updown-bar.png",
       plot   = bp6,
       dpi    = 600,
       device = "png",
       width  = 6,
       height = 5,
       units  = "in")

ggsave("DESEQ2_output/triploid_desiccation/triploid_desiccation-DEG-apeglm.png",
       plot   = bp7,
       dpi    = 600,
       device = "png",
       width  = 5,
       height = 4,
       units  = "in")

ggsave("DESEQ2_output/triploid_desiccation/triploid_desiccation-volcano_pval_genes_apeglm.png",
       plot   = p9,
       dpi    = 600,
       device = "png",
       width  = 4,
       height = 4,
       units  = "in")


beep(2)
```

```{r, include = TRUE}
beep(2); Sys.sleep(0.3); beep(2); Sys.sleep(0.3); beep(2); Sys.sleep(0.3); beep(4)
```

# Generate table for IPA
```{R eval=FALSE, include=FALSE}
rm(list=ls())

control_ploidy     <- read.table("DESEQ2_output/control_ploidy/CONTROL_PLOIDY-ALL-DEG-apeglm.csv", header = TRUE)
heat_ploidy        <- read.table("DESEQ2_output/heat_ploidy/HEAT_PLOIDY-ALL-DEG-apeglm.csv", header = TRUE)
desiccation_ploidy <- read.table("DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-ALL-DEG-apeglm.csv", header = TRUE)
diploid_heat          <- read.table("DESEQ2_output/diploid_heat/DIPLOID_HEAT-ALL-DEG-apeglm.csv", header = TRUE)
diploid_desciccation  <- read.table("DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-ALL-DEG-apeglm.csv", header = TRUE)
triploid_heat         <- read.table("DESEQ2_output/triploid_heat/TRIPLOID_HEAT-ALL-DEG-apeglm.csv", header = TRUE)
triploid_desiccation  <- read.table("DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-ALL-DEG-apeglm.csv", header = TRUE)

control_ploidy       <- control_ploidy        %>% dplyr::select(treatment,gene,padj,log2FoldChange)
heat_ploidy          <- heat_ploidy           %>% dplyr::select(treatment,gene,padj,log2FoldChange)
desiccation_ploidy   <- desiccation_ploidy    %>% dplyr::select(treatment,gene,padj,log2FoldChange)
diploid_heat         <- diploid_heat          %>% dplyr::select(treatment,gene,padj,log2FoldChange)
diploid_desciccation <- diploid_desciccation  %>% dplyr::select(treatment,gene,padj,log2FoldChange)
triploid_heat        <- triploid_heat         %>% dplyr::select(treatment,gene,padj,log2FoldChange)
triploid_desiccation <- triploid_desiccation  %>% dplyr::select(treatment,gene,padj,log2FoldChange)

all_trt <- rbind(control_ploidy,
                 heat_ploidy,
                 desiccation_ploidy,
                 diploid_heat,
                 diploid_desciccation,
                 triploid_heat,
                 triploid_desiccation)

# check row sums match
sum(nrow(control_ploidy),nrow(heat_ploidy),nrow(desiccation_ploidy ),
    nrow(diploid_heat ),nrow(diploid_desciccation),nrow(triploid_heat ),
    nrow(triploid_desiccation))
nrow(all_trt)

write.table(all_trt, "DESEQ2_output/2022_cgigas_DEG-all-treatments.txt", row.names = FALSE)

```

